From ba7b7566e8df885540e1c31008c2ce5614f0b6a7 Mon Sep 17 00:00:00 2001
From: Arnon Kaufman <akaufman@habana.ai>
Date: Wed, 01 May 2024 09:29:20 +0300
Subject: [PATCH] [SW-15774] board: accepting full path

Prepare functions to accept full path as parameters
Make use of the absolute file locations.
Prepare the scripts to use full paths

HL_DEV may be <board> or <board>_arc. DEV_PREFIX will be <board>
for both cases. Therefore ${HL_DEV}_arc can be extract to
<board>_arc_arc.

Change-Id: I77452d48bbce63bf016302c7cdeafb0c88b97aa2
Signed-off-by: Arnon Kaufman <akaufman@habana.ai>
---

diff --git a/board/habanalabs/common/create-eeprom-fit.sh b/board/habanalabs/common/create-eeprom-fit.sh
index f8a0d01..8dc56a3 100755
--- a/board/habanalabs/common/create-eeprom-fit.sh
+++ b/board/habanalabs/common/create-eeprom-fit.sh
@@ -11,17 +11,15 @@
 ADD_FILE_SUFFIX=${3:+-$3}	# add '-' (dash) only if supplied with 3rd argument
 BOARD_BASE=${BR2_EXTERNAL}/board/habanalabs/${HL_DEV}
 MKIMAGE_FILE=${BR2_EXTERNAL}/output/${HL_DEV}/host/bin/mkimage
-KEY_DIR=${BR2_EXTERNAL}/board/habanalabs/common/key/u-boot
-IMAGE_DIRETORY=${BR2_EXTERNAL}/output/${HL_DEV}/images
 KEY_FILE_NAME=hl_rsa3072.key
 KEY_FILE_DTB=hl_rsa3072.dtb
 EEPROM_ITS_TEMPLATE=${BR2_EXTERNAL}/board/habanalabs/common/eeprom.its
 FIT_LOAD_TYPE_EEPROM=$(grep -w FIT_LOAD_TYPE_EEPROM ${SPECS_EMB_PROJ_DIR}/fit_load_type.h | sed 's/.* = \(.*\),/\1/')
-EEPROM_ITS_FILE=$IMAGE_DIRETORY/${DEV_PREFIX}-eeprom${ADD_FILE_SUFFIX}.its
-EEPROM_ITB_FILE=$IMAGE_DIRETORY/${DEV_PREFIX}-eeprom${ADD_FILE_SUFFIX}.itb
+EEPROM_ITS_FILE=${STAGE_DIR}/${DEV_PREFIX}-eeprom${ADD_FILE_SUFFIX}.its
+EEPROM_ITB_FILE=${ITBS_DIR}/${DEV_PREFIX}-eeprom${ADD_FILE_SUFFIX}.itb
 KEYS_DIR=${BR2_EXTERNAL}/board/habanalabs/${HL_DEV}/key/
 EEPROM_IMAGE_NAME=eeprom${ADD_FILE_SUFFIX}.txt
-EEPROM_IMAGE=${IMAGE_DIRETORY}/${EEPROM_IMAGE_NAME}
+EEPROM_IMAGE=${IMGS_DIR}/${EEPROM_IMAGE_NAME}
 # Set the security key index and version
 HL_SEC_FILE=${SPECS_EMB_PROJ_DIR}/${DEV_PREFIX}/${DEV_PREFIX}_security.h
 HL_SEC_KEY_INDEX=`grep HL_SEC_KEY_INDEX $HL_SEC_FILE | sed 's/^.*\(0x.*\)$/\1/'`
@@ -64,7 +62,7 @@
 	mv ${TMP_FILE} ${1}
 }
 
-pushd ${IMAGE_DIRETORY}
+pushd ${STAGE_DIR}
 	eeprom_check_input
 	cp ${KEYS_DIR}/${KEY_FILE_NAME} .
 	sed ${ITS_SED_PARAMS} ${EEPROM_ITS_TEMPLATE} > ${EEPROM_ITS_FILE}
diff --git a/board/habanalabs/common/dirs.sh b/board/habanalabs/common/dirs.sh
index bad3f9c..12cc9a0 100644
--- a/board/habanalabs/common/dirs.sh
+++ b/board/habanalabs/common/dirs.sh
@@ -1,7 +1,7 @@
 # NOTE: should be in sync with scripts/make/common/hl_dirs.mk
 
 export DEV_BASE_DIR=${BR2_EXTERNAL}/output/${HL_DEV}/
-export DEV_PREFIX_BASE_DIR=${BR2_EXTERNAL}/output/${DEV_PREFIX}/
+export DEV_PREFIX_BASE_DIR=${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/
 export IMGS_DIR=${DEV_BASE_DIR}/images/
 export IMGS_PROD_DIR=${DEV_PREFIX_BASE_DIR}/images/
 export STAGE_DIR=${IMGS_DIR}
diff --git a/board/habanalabs/common/post-image-agents.sh b/board/habanalabs/common/post-image-agents.sh
index 3930d29..952f11b 100755
--- a/board/habanalabs/common/post-image-agents.sh
+++ b/board/habanalabs/common/post-image-agents.sh
@@ -57,10 +57,10 @@
 
 	# copy img-pointer, ppboot, preboot, security binaries from other HL targets required for agent fit creation
 	# This will be called by gaudi2 agent scripts
-	cp ${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/images/${HL_DEV}-ppboot-pcie_phy_fw.bin.le ${STAGE_DIR}/${HL_DEV}-ppboot-arc.bin.le
-	cp ${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/images/${HL_DEV}-preboot.bin.le ${STAGE_DIR}/${HL_DEV}-preboot-arc.bin.le
-	cp ${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/images/${HL_DEV}-images-pointers.bin.le ${STAGE_DIR}/${HL_DEV}-images-pointers-arc.bin.le
-	cp ${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/images/${HL_DEV}-security-arc.bin.le ${STAGE_DIR}
+	cp ${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/images/${HL_DEV}-ppboot-pcie_phy_fw.bin.le ${BINS_DIR}/${HL_DEV}-ppboot-arc.bin.le
+	cp ${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/images/${HL_DEV}-preboot.bin.le ${BINS_DIR}/${HL_DEV}-preboot-arc.bin.le
+	cp ${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/images/${HL_DEV}-images-pointers.bin.le ${BINS_DIR}/${HL_DEV}-images-pointers-arc.bin.le
+	cp ${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/images/${HL_DEV}-security-arc.bin.le ${BINS_DIR}
 }
 
 function handle_agent_fit {
@@ -71,13 +71,13 @@
 
 	AGENT_ITS_TEMPLATE=${ITS_PATH}/agent-${AGENT_NAME}_${IMAGE_TYPE}.its
 
-	sed ${ITS_SED_PARAMS}${ITS_SED_BIN_PARAMS} ${AGENT_ITS_TEMPLATE} > ${AGENT_ITS_FILE}
-	${HOST_DIR}/usr/bin/mkimage -D "-W no-unit_address_vs_reg" -f ${AGENT_ITS_FILE} -k ${KEYS_DIR} ${AGENT_ITB_FILE}
+	sed ${ITS_SED_PARAMS}${ITS_SED_BIN_PARAMS} ${AGENT_ITS_TEMPLATE} > ${STAGE_DIR}/${AGENT_ITS_FILE}
+	${HOST_DIR}/usr/bin/mkimage -D "-W no-unit_address_vs_reg" -f ${STAGE_DIR}/${AGENT_ITS_FILE} -k ${KEYS_DIR} ${STAGE_DIR}/${AGENT_ITB_FILE}
 
 	# KMD like this to be 4B aligned in size
-	dd status=none conv=sync bs=4 if=${AGENT_ITB_FILE} > ${DEV_PREFIX}-${AGENT_ITB_FILE}
+	dd status=none conv=sync bs=4 if=${STAGE_DIR}/${AGENT_ITB_FILE} > ${ITBS_DIR}/${DEV_PREFIX}-${AGENT_ITB_FILE}
 
-	rm -f ${AGENT_ITB_FILE}
+	rm -f ${STAGE_DIR}/${AGENT_ITB_FILE}
 }
 
 pushd ${STAGE_DIR}
@@ -140,18 +140,18 @@
 		AGENTS_ITS_SPEC_TEMPLATE=${BR2_EXTERNAL}/board/habanalabs/${HL_DEV}/agent-${AGENT_NAME}.its
 		AGENTS_ITS_SPEC_FILE=agent-${AGENT_NAME}-${HL_DEV}.its
 		if [ -f ${AGENTS_ITS_SPEC_TEMPLATE} ]; then
-			sed ${ITS_SED_PARAMS}${ITS_SED_BIN_PARAMS} ${AGENTS_ITS_SPEC_TEMPLATE} > ${AGENTS_ITS_SPEC_FILE}
-			ITS_SED_BIN_PARAMS+="s#@DEV_SPEC_ITS@#${PWD}/${AGENTS_ITS_SPEC_FILE}#g;"
+			sed ${ITS_SED_PARAMS}${ITS_SED_BIN_PARAMS} ${AGENTS_ITS_SPEC_TEMPLATE} > ${STAGE_DIR}/${AGENTS_ITS_SPEC_FILE}
+			ITS_SED_BIN_PARAMS+="s#@DEV_SPEC_ITS@#${STAGE_DIR}/${AGENTS_ITS_SPEC_FILE}#g;"
 		fi
 
-		sed ${ITS_SED_PARAMS}${ITS_SED_BIN_PARAMS} ${AGENT_ITS_TEMPLATE} > ${AGENT_ITS_FILE}
-		${GCC_ARC} -nostdinc -I include -undef -x assembler-with-cpp -P -E ${AGENT_ITS_FILE} > ${AGENT_ITS_FILE}.post
-		${HOST_DIR}/usr/bin/mkimage -D "-W no-unit_address_vs_reg" -f ${AGENT_ITS_FILE}.post -k ${KEYS_DIR} ${AGENT_ITB_FILE}
+		sed ${ITS_SED_PARAMS}${ITS_SED_BIN_PARAMS} ${AGENT_ITS_TEMPLATE} > ${STAGE_DIR}/${AGENT_ITS_FILE}
+		${GCC_ARC} -nostdinc -I include -undef -x assembler-with-cpp -P -E ${STAGE_DIR}/${AGENT_ITS_FILE} > ${STAGE_DIR}/${AGENT_ITS_FILE}.post
+		${HOST_DIR}/usr/bin/mkimage -D "-W no-unit_address_vs_reg" -f ${STAGE_DIR}/${AGENT_ITS_FILE}.post -k ${KEYS_DIR} ${STAGE_DIR}/${AGENT_ITB_FILE}
 
 		# KMD like this to be 4B aligned in size
-		dd status=none conv=sync bs=4 if=${AGENT_ITB_FILE} > ${HL_DEV}-${AGENT_ITB_FILE}
+		dd status=none conv=sync bs=4 if=${STAGE_DIR}/${AGENT_ITB_FILE} > ${ITBS_DIR}/${HL_DEV}-${AGENT_ITB_FILE}
 
-		rm -f ${AGENT_ITB_FILE}
+		rm -f ${STAGE_DIR}/${AGENT_ITB_FILE}
 
 		# device-specific post-image work
 		post_image_agent_handle
diff --git a/board/habanalabs/common/post-image-bins.sh b/board/habanalabs/common/post-image-bins.sh
index 0f83d2c..8cf4f07 100755
--- a/board/habanalabs/common/post-image-bins.sh
+++ b/board/habanalabs/common/post-image-bins.sh
@@ -16,7 +16,7 @@
 	local LEFILENAME=${FILENAMEPATH}/${HL_DEV}-${FILENAME}.le
 	local BEFILENAME=${FILENAMEPATH}/${HL_DEV}-${FILENAME}.be
 
-	dd status=none conv=sync bs=${BS} if=${FILENAME} of=${LEFILENAME}
+	dd status=none conv=sync bs=${BS} if=${FULLFILENAME} of=${LEFILENAME}
 	swap_file ${LEFILENAME} ${BEFILENAME}
 	hl_compress_bin ${LEFILENAME}
 }
diff --git a/board/habanalabs/common/post-image-exports.sh b/board/habanalabs/common/post-image-exports.sh
index 47db09a..1c1e0d5 100755
--- a/board/habanalabs/common/post-image-exports.sh
+++ b/board/habanalabs/common/post-image-exports.sh
@@ -1,7 +1,7 @@
 #!/usr/bin/env bash
 
 # Set U-Boot binary file name for .its
-HL_UBOOT_BIN_FILE=${HL_DEV}-u-boot.bin.le
+HL_UBOOT_BIN_FILE=${BINS_DIR}/${HL_DEV}-u-boot.bin.le
 
 ITS_SED_PARAMS+="s/@HL_VERSION@/${HL_VERSION}/g;"
 ITS_SED_PARAMS+="s/@HL_DEV@/${HL_DEV}/g;"
diff --git a/board/habanalabs/common/post-image-fit-arc.sh b/board/habanalabs/common/post-image-fit-arc.sh
index b892eb0..6bf0345 100755
--- a/board/habanalabs/common/post-image-fit-arc.sh
+++ b/board/habanalabs/common/post-image-fit-arc.sh
@@ -12,8 +12,8 @@
 HL_VERSION=`cut -d- -f4 ${BR2_EXTERNAL}/output/${HL_DEV}/build/zephyr-custom/localversion-habanalabs`
 
 COMMON_BASE=${BR2_EXTERNAL}/board/habanalabs/common
-ZEPHYR_ITS_FILE=zephyr-${DEV_NAME}_${HL_TARGET}.its
-ZEPHYR_ITB_FILE=zephyr-fit.itb
+ZEPHYR_ITS_FILE=${STAGE_DIR}/zephyr-${DEV_NAME}_${HL_TARGET}.its
+ZEPHYR_ITB_FILE=${STAGE_DIR}/zephyr-fit.itb
 
 # Set compression to lzo
 HL_COMPRESS=lzo
@@ -68,12 +68,12 @@
 
 pushd ${STAGE_DIR}
 
-	hl_copy_keys ${STAGE_DIR}
+	hl_copy_keys ${BINS_DIR}
 
 	hl_copy_binaries ${BINS_DIR}
 
 	sed ${ITS_SED_PARAMS} ${ZEPHYR_ITS_TEMPLATE} > ${ZEPHYR_ITS_FILE}
 	${MKIMAGE} -D "-W no-unit_address_vs_reg" -f ${ZEPHYR_ITS_FILE} -k ${STAGE_DIR} ${ZEPHYR_ITB_FILE}
 
-	dd status=none conv=sync bs=4 if=${ZEPHYR_ITB_FILE} > ${DEV_PREFIX}-boot-fit.itb
+	dd status=none conv=sync bs=4 if=${ZEPHYR_ITB_FILE} > ${ITBS_DIR}/${DEV_PREFIX}-boot-fit.itb
 popd
diff --git a/board/habanalabs/common/post-image-fit-descriptor.sh b/board/habanalabs/common/post-image-fit-descriptor.sh
index caaaafb..e5bdfc0 100755
--- a/board/habanalabs/common/post-image-fit-descriptor.sh
+++ b/board/habanalabs/common/post-image-fit-descriptor.sh
@@ -31,7 +31,7 @@
 
 pushd ${BIN_IMG_DIR}
 	hl_handle_desc
-	AGENT_FIT_FILES=$(ls ${HL_DEV}-agent-*.itb 2>/dev/null)
+	AGENT_FIT_FILES=$(ls ${ITBS_DIR}/${HL_DEV}-agent-*.itb 2>/dev/null)
 
 	# For each agent-*.itb files
 	for AGENT_FIT in $AGENT_FIT_FILES; do
diff --git a/board/habanalabs/common/post-image-fit-global.sh b/board/habanalabs/common/post-image-fit-global.sh
index f58e455..07792c0 100755
--- a/board/habanalabs/common/post-image-fit-global.sh
+++ b/board/habanalabs/common/post-image-fit-global.sh
@@ -40,22 +40,26 @@
 HL_ZEPHYR_MGMT_VERSION=`cat ${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/build/zephyr-custom/build-mgmt/buildversion-habanalabs`
 HL_ZEPHYR_PID_VERSION=`cat ${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/build/zephyr-custom/build-pid/buildversion-habanalabs`
 
-UBOOT_ZEPHYR_ITS_FILE=uboot-zephyr-${DEV_NAME}_${HL_TARGET}.its
-UBOOT_ZEPHYR_ITB_FILE=uboot-zephyr-fit.itb
+UBOOT_ZEPHYR_ITS_FILE=${STAGE_DIR}/uboot-zephyr-${DEV_NAME}_${HL_TARGET}.its
+UBOOT_ZEPHYR_ITB_FILE=${STAGE_DIR}/uboot-zephyr-fit.itb
 NIC_FW_FILE=${BR2_EXTERNAL}/board/habanalabs/${HL_DEV}/fw/nic_fw.bin
 
-SPI_ITS_FILE=spi-${DEV_NAME}_${HL_TARGET}.its
-SPI_ITB_FILE=spi-fpga-fit.itb
+SPI_ITS_FILE=${STAGE_DIR}/spi-${DEV_NAME}_${HL_TARGET}.its
+SPI_ITB_FILE=${STAGE_DIR}/spi-fpga-fit.itb
 SPI_FPGA_ITS_TEMPLATE=${COMMON_BASE}/spi-fpga-secureboot.its
 
-LINUX_ITS_FILE=linux-${DEV_NAME}_${HL_TARGET}.its
-LINUX_ITB_FILE=linux-fit.itb
+LINUX_ITS_FILE=${STAGE_DIR}/linux-${DEV_NAME}_${HL_TARGET}.its
+LINUX_ITB_FILE=${STAGE_DIR}/linux-fit.itb
 LINUX_ITS_TEMPLATE=${COMMON_BASE}/linux.its
 
-DSV_BIN=dsv.bin
-DSV_ITS_FILE=dsv.its
-DSV_ITS_TEMPLATE=${COMMON_BASE}/${DSV_ITS_FILE}
-DSV_ITB_FILE=dsv-fit.itb
+DSV_BIN=${BINS_DIR}/dsv.bin
+DSV_ITS_FILE=${STAGE_DIR}/dsv.its
+DSV_ITS_TEMPLATE=${COMMON_BASE}/dsv.its
+DSV_ITB_FILE=${ITBS_DIR}dsv-fit.itb
+
+LINUX_PRODUCTS_DIR=${DEV_BASE_DIR}/build/linux-custom/arch/arm64/boot/
+LINUX_DTB_DIR=${LINUX_PRODUCTS_DIR}/dts/habanalabs/${HL_DEV}/
+LINUX_ROOTFS_DIR=${IMGS_DIR}
 
 LINUX_PRODUCTS_DIR=${DEV_BASE_DIR}/build/linux-custom/arch/arm64/boot/
 LINUX_DTB_DIR=${LINUX_PRODUCTS_DIR}/dts/habanalabs/${HL_DEV}/
@@ -137,20 +141,20 @@
 ITS_SED_PARAMS+="s#@LINUX_ROOTFS_DIR@#${LINUX_ROOTFS_DIR}#g;"
 
 if [ ! -f ${NIC_FW_FILE} ]; then
-	dd bs=4 count=1 if=/dev/zero of=${STAGE_DIR}/empty_nic_fw.bin
-	NIC_FW_FILE=${STAGE_DIR}/empty_nic_fw.bin
+	dd bs=4 count=1 if=/dev/zero of=${BINS_DIR}/empty_nic_fw.bin
+	NIC_FW_FILE=${BINS_DIR}/empty_nic_fw.bin
 fi
 ITS_SED_PARAMS+="s#@NIC_FW@#${NIC_FW_FILE}#g;"
 
 function generate_dsv_fit
 {
-	if [ ! -f ${STAGE_DIR}/${DSV_BIN} ]; then
+	if [ ! -f ${DSV_BIN} ]; then
 		return
 	fi
 
 	# Compress binary
 	if [ ${HL_COMPRESS} == "lzo" ]; then
-		${LZOP} -f ${STAGE_DIR}/${DSV_BIN} -o ${STAGE_DIR}/${DSV_BIN}.lzop
+		${LZOP} -f ${DSV_BIN} -o ${DSV_BIN}.lzop
 		DSV_BIN+=.lzop
 	elif [ ${HL_COMPRESS} != "none" ]; then
 		echo "Unsupported compression ${HL_COMPRESS}" >&2
@@ -162,14 +166,14 @@
 	${MKIMAGE} -D "-W no-unit_address_vs_reg" -f ${DSV_ITS_FILE} ${DSV_ITB_FILE}
 
 	# LKD likes this to be 4B aligned in size
-	dd status=none conv=sync bs=4 if=${DSV_ITB_FILE} > ${DEV_PREFIX}-${DSV_ITB_FILE}
+	dd status=none conv=sync bs=4 if=${DSV_ITB_FILE} > ${ITBS_DIR}/${DEV_PREFIX}-${DSV_ITB_FILE}
 
 	rm -f ${DSV_ITB_FILE}
 }
 
 pushd ${STAGE_DIR}
 
-	hl_copy_keys ${STAGE_DIR}
+	hl_copy_keys ${BINS_DIR}
 
 	hl_copy_binaries ${BINS_DIR}
 
@@ -179,7 +183,7 @@
 	if [ ${HL_TARGET} == "onix" ]; then
 		sed ${ITS_SED_PARAMS} ${SPI_FPGA_ITS_TEMPLATE} > ${SPI_ITS_FILE}
 		${MKIMAGE} -D "-W no-unit_address_vs_reg" -f ${SPI_ITS_FILE} ${SPI_ITB_FILE}
-		dd status=none conv=sync bs=4 if=${SPI_ITB_FILE} > ${DEV_NAME}-boot-spi-fpga-fit.itb
+		dd status=none conv=sync bs=4 if=${SPI_ITB_FILE} > ${ITBS_DIR}/${DEV_NAME}-boot-spi-fpga-fit.itb
 	fi
 
 	uboot_bin_sz=$(stat -Lc "%s" ${HL_UBOOT_BIN_FILE})
@@ -188,15 +192,15 @@
 		exit 1
 	fi
 	sed ${ITS_SED_PARAMS} ${UBOOT_ZEPHYR_ITS_TEMPLATE} > ${UBOOT_ZEPHYR_ITS_FILE}
-	${MKIMAGE} -D "-W no-unit_address_vs_reg" -f ${UBOOT_ZEPHYR_ITS_FILE} -k ${STAGE_DIR} ${UBOOT_ZEPHYR_ITB_FILE}
+	${MKIMAGE} -D "-W no-unit_address_vs_reg" -f ${UBOOT_ZEPHYR_ITS_FILE} -k ${BINS_DIR} ${UBOOT_ZEPHYR_ITB_FILE}
 
 	# TODO: Replace DEV_PREFIX to DEV_NAME
-	dd status=none conv=sync bs=4 if=${UBOOT_ZEPHYR_ITB_FILE} > ${DEV_PREFIX}-boot-fit.itb
+	dd status=none conv=sync bs=4 if=${UBOOT_ZEPHYR_ITB_FILE} > ${ITBS_DIR}/${DEV_PREFIX}-boot-fit.itb
 
 	sed ${ITS_SED_PARAMS} ${LINUX_ITS_TEMPLATE} > ${LINUX_ITS_FILE}
-	${MKIMAGE} -D "-W no-unit_address_vs_reg" -f ${LINUX_ITS_FILE} -k ${STAGE_DIR} ${LINUX_ITB_FILE}
+	${MKIMAGE} -D "-W no-unit_address_vs_reg" -f ${LINUX_ITS_FILE} -k ${BINS_DIR} ${LINUX_ITB_FILE}
 
-	dd status=none conv=sync bs=4 if=${LINUX_ITB_FILE} > ${DEV_PREFIX}-fit.itb
+	dd status=none conv=sync bs=4 if=${LINUX_ITB_FILE} > ${ITBS_DIR}/${DEV_PREFIX}-fit.itb
 
 	generate_dsv_fit
 popd
diff --git a/board/habanalabs/common/post-image-fit.sh b/board/habanalabs/common/post-image-fit.sh
index 0255dd7..80e19c5 100755
--- a/board/habanalabs/common/post-image-fit.sh
+++ b/board/habanalabs/common/post-image-fit.sh
@@ -26,12 +26,12 @@
 IFS=$'\n'
 
 # Set hl-version = "x.y.z"; inside FIT header to be used for KMD/FW compatability validation
-HL_VERSION=`cut -d- -f4 ${BASE_DIR}/build/linux-custom/localversion-habanalabs`
+HL_VERSION=`cut -d- -f4 ${DEV_BASE_DIR}/build/linux-custom/localversion-habanalabs`
 # Set hl-*-version = "hl-<device>-<HL_VERSION>[-rc]-fw-<fw ver>[-<git data>]"; inside FIT header
-HL_FW_VERSION=`cut -d- -f2- ${BASE_DIR}/build/linux-custom/localversion-habanalabs`
-HL_UBOOT_VERSION=`cut -d- -f2- ${BASE_DIR}/build/uboot-custom/localversion-habanalabs`
+HL_FW_VERSION=`cut -d- -f2- ${DEV_BASE_DIR}/build/linux-custom/localversion-habanalabs`
+HL_UBOOT_VERSION=`cut -d- -f2- ${DEV_BASE_DIR}/build/uboot-custom/localversion-habanalabs`
 # Set hl-preboot-version = "hl-<device>-<HL_VERSION>[-rc]-fw-<fw ver>[-<git data>] (<date> - <time>)"
-HL_PREBOOT_VERSION=`cat ${BASE_DIR}/build/baremetal-custom/buildversion-habanalabs`
+HL_PREBOOT_VERSION=`cat ${DEV_BASE_DIR}/build/baremetal-custom/buildversion-habanalabs`
 # Set enum number of FIT_LOAD_TYPE_FLASH in FIT
 FIT_LOAD_TYPE_FLASH=$(grep -E "FIT_LOAD_TYPE_FLASH[^_]" ${SPECS_EMBEDDED_DIR}/fit_load_type.h | sed 's/.* = \(.*\),/\1/')
 # Set enum number of FIT_LOAD_TYPE_RUN in FIT
@@ -42,14 +42,14 @@
 HL_SEC_FW_VERSION=`grep HL_SEC_FW_VERSION $HL_SEC_FILE | awk '{ print $NF }'`
 
 NIC_FW_FILE=${BR2_EXTERNAL_HABANALABS_PATH}/board/habanalabs/${HL_DEV}/fw/condor.fw.bin
-UBOOT_ITS_FILE=uboot-${HL_DEV}_${HL_TARGET}.its
-UBOOT_ITB_FILE=uboot-fit.itb
+UBOOT_ITS_FILE=${STAGE_DIR}/uboot-${HL_DEV}_${HL_TARGET}.its
+UBOOT_ITB_FILE=${STAGE_DIR}/uboot-fit.itb
 UBOOT_ITS_TEMPLATE=`sed -e 's/^"//' -e 's/"$//' <<< ${BR2_POST_IMAGE_UBOOT_ITS_TEMPLATE}`
-SPI_ITS_FILE=spi-${HL_DEV}_${HL_TARGET}.its
-SPI_ITB_FILE=spi-fpga-fit.itb
+SPI_ITS_FILE=${STAGE_DIR}/spi-${HL_DEV}_${HL_TARGET}.its
+SPI_ITB_FILE=${STAGE_DIR}/spi-fpga-fit.itb
 SPI_FPGA_ITS_TEMPLATE=${COMMON_BASE}/spi-fpga-legacy.its
-LINUX_ITS_FILE=linux-${HL_DEV}_${HL_TARGET}.its
-LINUX_ITB_FILE=linux-fit.itb
+LINUX_ITS_FILE=${STAGE_DIR}/linux-${HL_DEV}_${HL_TARGET}.its
+LINUX_ITB_FILE=${STAGE_DIR}/linux-fit.itb
 BOARD_BASE=${BR2_EXTERNAL_HABANALABS_PATH}/board/habanalabs/${HL_DEV}
 
 if [ -e ${COMMON_BASE}/linux-${HL_TARGET}.its ]; then
@@ -62,10 +62,10 @@
 LINUX_DTB_DIR=${LINUX_PRODUCTS_DIR}/dts/habanalabs/${HL_DEV}/
 LINUX_ROOTFS_DIR=${IMGS_DIR}
 
-DSV_BIN=dsv.bin
-DSV_ITS_FILE=dsv.its
+DSV_BIN=${BINS_DIR}/dsv.bin
+DSV_ITS_FILE=${STAGE_DIR}/dsv.its
 DSV_ITS_TEMPLATE=${COMMON_BASE}/${DSV_ITS_FILE}
-DSV_ITB_FILE=dsv-fit.itb
+DSV_ITB_FILE=${STAGE_DIR}/dsv-fit.itb
 
 source ${BOARD_BASE}/post-image-fit.sh
 source ${BOARD_BASE}/post-image-common.sh
@@ -89,7 +89,7 @@
 
 function generate_dsv_fit
 {
-	if [ ! -f ${STAGE_DIR}/${DSV_BIN} ]; then
+	if [ ! -f ${DSV_BIN} ]; then
 		return
 	fi
 
@@ -104,7 +104,7 @@
 		sed ${ITS_SED_PARAMS} ${SPI_FPGA_ITS_TEMPLATE} > ${SPI_ITS_FILE}
 		${HOST_DIR}/usr/bin/mkimage -D "-W no-unit_address_vs_reg" -f ${SPI_ITS_FILE} -k ${KEYS_DIR} ${SPI_ITB_FILE}
 		# LKD like this to be 4B aligned in size
-		dd status=none conv=sync bs=4 if=${SPI_ITB_FILE} > ${HL_DEV}-boot-spi-fpga-fit.itb
+		dd status=none conv=sync bs=4 if=${SPI_ITB_FILE} > ${ITBS_DIR}/${HL_DEV}-boot-spi-fpga-fit.itb
 	fi
 
 	if [ ! -f ${NIC_FW_FILE} ]; then
@@ -122,14 +122,14 @@
 	${HOST_DIR}/usr/bin/mkimage -D "-W no-unit_address_vs_reg" -f ${UBOOT_ITS_FILE} -k ${KEYS_DIR} ${UBOOT_ITB_FILE}
 
 	# LKD like this to be 4B aligned in size
-	dd status=none conv=sync bs=4 if=${UBOOT_ITB_FILE} > ${HL_DEV}-boot-fit.itb
+	dd status=none conv=sync bs=4 if=${UBOOT_ITB_FILE} > ${ITBS_DIR}/${HL_DEV}-boot-fit.itb
 
 	sed ${ITS_SED_PARAMS} ${LINUX_ITS_TEMPLATE} > ${LINUX_ITS_FILE}
-	cp ${BASE_DIR}/build/linux-custom/arch/arm64/boot/Image .
+	cp ${LINUX_PRODUCTS_DIR}/Image ${BINS_DIR}
 	${HOST_DIR}/usr/bin/mkimage -D "-W no-unit_address_vs_reg" -f ${LINUX_ITS_FILE} -k ${KEYS_DIR} ${LINUX_ITB_FILE}
 
 	# KMD like this to be 4B aligned in size
-	dd status=none conv=sync bs=4 if=${LINUX_ITB_FILE} > ${HL_DEV}-fit.itb
+	dd status=none conv=sync bs=4 if=${LINUX_ITB_FILE} > ${ITBS_DIR}/${HL_DEV}-fit.itb
 
 	generate_dsv_fit
 popd
diff --git a/board/habanalabs/common/post-image-images-pointers.sh b/board/habanalabs/common/post-image-images-pointers.sh
index 8f00970..f527141 100755
--- a/board/habanalabs/common/post-image-images-pointers.sh
+++ b/board/habanalabs/common/post-image-images-pointers.sh
@@ -87,8 +87,8 @@
 	local img_desc_sz=44
 
 	# Prepare the NEXT and MODIFIABLE for backward compatibility
-	echo "11110000" > ${HL_DEV}-next-image.txt
-	dd status=none conv=sync bs=8 count=1 if=${HL_DEV}-next-image.txt of=$img_file
+	echo "11110000" > ${STAGE_DIR}/${HL_DEV}-next-image.txt
+	dd status=none conv=sync bs=8 count=1 if=${STAGE_DIR}/${HL_DEV}-next-image.txt of=$img_file
 
 	# Pointers structure magic "NXT"0x0
 	printf "%x%x%x00" "'N" "'X" "'T" | rev_bytes_stdin | xxd -r -p -s +8 - $img_file
diff --git a/board/habanalabs/common/post-image-spi-arc.sh b/board/habanalabs/common/post-image-spi-arc.sh
index cc33e7a..36cc0a9 100755
--- a/board/habanalabs/common/post-image-spi-arc.sh
+++ b/board/habanalabs/common/post-image-spi-arc.sh
@@ -6,6 +6,6 @@
 pushd ${STAGE_DIR}
 
 	# for SecureBoot ROM
-	hl_pldm_hex arcboot_rom arcboot_rom.bin 16
+	hl_pldm_hex arcboot_rom arcboot_rom.bin 16 ${HEXS_DIR}
 
 popd
diff --git a/board/habanalabs/common/post-image-spi-arc_global.sh b/board/habanalabs/common/post-image-spi-arc_global.sh
index 1a87551..e596afd 100755
--- a/board/habanalabs/common/post-image-spi-arc_global.sh
+++ b/board/habanalabs/common/post-image-spi-arc_global.sh
@@ -148,7 +148,7 @@
 	FILE=$1
 	IMAGE_NUM=$2
 	hl_get_size ${1} FLASH_BASE_ADDR
-	dd status=none conv=notrunc seek=${IMAGE_NUM} bs=4 if=${FILE}.size of=${DEV_PREFIX}-spi-flash.bin
+	dd status=none conv=notrunc seek=${IMAGE_NUM} bs=4 if=${FILE}.size of=${BINS_DIR}/${DEV_PREFIX}-spi-flash.bin
 	rm -rf ${FILE}.size
 }
 
@@ -175,7 +175,7 @@
 		for i in {0..3}; do
 			DWORD=$(echo ${VERSION} | xxd -r -p | xxd -g4 | cut -d' ' -f$(( 5 - ${i} )))
 			FORCE="{h5_core.h5_psoc_top.psoc_macro.efuse.efuse_cfg_reg.u_efuse_cfg_reg_fields.u_cfg_reg_base_data_$(( 59 + ${i}))_r.data_in_r[31:0]}"
-			echo "force ${FORCE} 'h${DWORD}" >> ${HL_TARGET}-force.hex
+			echo "force ${FORCE} 'h${DWORD}" >> ${HEXS_DIR}/${HL_TARGET}-force.hex
 		done
 	fi
 }
@@ -195,7 +195,7 @@
 			DWORD=$(xxd -e -c32 -g4 ${NAME}_${KEY_BASE_NAME_PUB}.hash | cut -d' ' -f$(( ${i} + 2 )))
 			OFFSET=$(( $(( ${KEY_OFFSET}*8 )) +  ${i} ))
 			FORCE="{h5_core.h5_psoc_top.psoc_macro.efuse.efuse_cfg_reg.u_efuse_cfg_reg_fields.u_cfg_reg_base_data_$(( 18 + ${OFFSET}))_r.data_in_r[31:0]}"
-			echo "force ${FORCE} 'h${DWORD}" >> ${HL_TARGET}-force.hex
+			echo "force ${FORCE} 'h${DWORD}" >> ${HEXS_DIR}/${HL_TARGET}-force.hex
 		done
 	fi
 }
@@ -203,9 +203,9 @@
 function hl_key_revoke {
 	REVOKE_STATUS=$1
 	if [ $REVOKE_STATUS -eq 1 ]; then
-		echo -n fe >> revoke_status.hex
+		echo -n fe >> ${STAGE_DIR}/revoke_status.hex
 	else
-		echo -n ff >> revoke_status.hex
+		echo -n ff >> ${STAGE_DIR}/revoke_status.hex
 	fi
 }
 
@@ -223,7 +223,7 @@
 	DWORD=$(( DWORD + $((${FW_VERSION_LOCATION}<<13))))
 	DWORD=$(( DWORD + $((${SPI_CONTROLLER}<<17))))
 	DWORD=$(printf "%08x" ${DWORD})
-	echo "force {h5_core.h5_psoc_top.psoc_macro.efuse.efuse_cfg_reg.u_efuse_cfg_reg_fields.u_cfg_reg_base_data_58_r.data_in_r[31:0]} 'h${DWORD}" > ${HL_TARGET}-force.hex
+	echo "force {h5_core.h5_psoc_top.psoc_macro.efuse.efuse_cfg_reg.u_efuse_cfg_reg_fields.u_cfg_reg_base_data_58_r.data_in_r[31:0]} 'h${DWORD}" > ${HEXS_DIR}/${HL_TARGET}-force.hex
 }
 
 function hl_set_revoke_status {
@@ -234,10 +234,10 @@
 	hl_key_revoke ${PROD3_KEY_REVOKE}
 	hl_key_revoke ${PROD4_KEY_REVOKE}
 
-	cat revoke_status.hex | xxd -r -p > ${FILE}.tmp
+	cat ${STAGE_DIR}/revoke_status.hex | xxd -r -p > ${FILE}.tmp
 	dd status=none conv=notrunc seek=11 bs=16 if=${FILE}.tmp of=${FILE}
 
-	rm -rf revoke_status.hex
+	rm -rf ${STAGE_DIR}/revoke_status.hex
 }
 
 function hl_update_image_offset {
@@ -416,10 +416,10 @@
 	done
 
 	## create otp spi image
-	dd status=none conv=sync count=12 bs=16 if=/dev/zero | tr '\000' $'\xFF' > ${DEV_PREFIX}-spi-otp-flash.bin
+	dd status=none conv=sync count=12 bs=16 if=/dev/zero | tr '\000' $'\xFF' > ${BINS_DIR}/${DEV_PREFIX}-spi-otp-flash.bin
 
 	if [ ${CREATE_DUMMY_FLASH} -eq 1 ]; then
-		dd status=none conv=sync count=${CREATE_DUMMY_FLASH_SIZE_64B_COUNT} bs=64 if=/dev/zero > ${DEV_PREFIX}-dummy.bin
+		dd status=none conv=sync count=${CREATE_DUMMY_FLASH_SIZE_64B_COUNT} bs=64 if=/dev/zero > ${BINS_DIR}/${DEV_PREFIX}-dummy.bin
 
 		PREBOOT_BINFILE=${DEV_PREFIX}-dummy.bin
 		PPBOOT_BINFILE=${DEV_PREFIX}-dummy.bin
@@ -434,22 +434,20 @@
 	generate_public_keys
 
 	# Create pif arc binaries
-	prep_image_pcie_pif ${PIF_ARC_ELFFILE}
-
+	prep_image_pcie_pif ${BINS_DIR}/${PIF_ARC_ELFFILE}
 	# Create erom binaries
-	prep_image_erom ${DEV_PREFIX}-erom.bin.le
-
+	prep_image_erom ${BINS_DIR}/${DEV_PREFIX}-erom.bin.le
 	# Create fake fuse binary
-	prep_image_fake_fuse ${DEV_PREFIX}-fake-fuse.bin.le
+	prep_image_fake_fuse ${BINS_DIR}/${DEV_PREFIX}-fake-fuse.bin.le
 
 	# Create sram mbist binary
-	prep_image_sram_mbist_data sram-mbist-data.bin
+	prep_image_sram_mbist_data ${BINS_DIR}/sram-mbist-data.bin
 
 	# Pad preboot binary and then pack
-	zephyr_preboot_pack ${PREBOOT_BINFILE} ${PRIME_IMAGE_RAM_KEY} ${PRIME_IMAGE_RAM_CRC_ERROR} ${PRIME_IMAGE_RAM_SIGN_ERROR}
+	zephyr_preboot_pack ${BINS_DIR}/${PREBOOT_BINFILE} ${PRIME_IMAGE_RAM_KEY} ${PRIME_IMAGE_RAM_CRC_ERROR} ${PRIME_IMAGE_RAM_SIGN_ERROR}
 
 	# Pad die1 preboot binary and then pack
-	zephyr_preboot_d1_pack ${PREBOOT_D1_BINFILE} ${PRIME_IMAGE_RAM_KEY} ${PRIME_IMAGE_RAM_CRC_ERROR} ${PRIME_IMAGE_RAM_SIGN_ERROR}
+	zephyr_preboot_d1_pack ${BINS_DIR}/${PREBOOT_D1_BINFILE} ${PRIME_IMAGE_RAM_KEY} ${PRIME_IMAGE_RAM_CRC_ERROR} ${PRIME_IMAGE_RAM_SIGN_ERROR}
 
 	## To be populated into FW SRAM
 
@@ -459,61 +457,61 @@
 	PPBOOT_PCIE_PHY_BINFILE=${PPBOOT}-${PCIE_PHY}.bin
 
 	# concatenate both FW
-	hl_join_bin ${PPBOOT} ${PCIE_PHY} ${PPBOOT_PCIE_PHY_BINFILE}
+	hl_join_bin ${BINS_DIR}/${PPBOOT} ${BINS_DIR}/${PCIE_PHY} ${BINS_DIR}/${PPBOOT_PCIE_PHY_BINFILE}
 
 	# pack and sign image
-	hl_pack_bin ${PPBOOT_PCIE_PHY_BINFILE} 64K ${PRIME_IMAGE_PPBOOT_KEY} ${PRIME_IMAGE_PPBOOT_CRC_ERROR} ${PRIME_IMAGE_PPBOOT_SIGN_ERROR}
+	hl_pack_bin ${BINS_DIR}/${PPBOOT_PCIE_PHY_BINFILE} 64K ${PRIME_IMAGE_PPBOOT_KEY} ${PRIME_IMAGE_PPBOOT_CRC_ERROR} ${PRIME_IMAGE_PPBOOT_SIGN_ERROR}
 
 	# Create image pointers
-	prep_image_ptrs ${DEV_PREFIX}-images-pointers.bin.le
+	prep_image_ptrs ${BINS_DIR}/${DEV_PREFIX}-images-pointers.bin.le
 
 	if [ ${HL_TARGET} == "pldm" ]; then
 		hl_efuse_sys_config_generate
 	fi
 
-	hl_sec_fw ${DEV_PREFIX}-spi-otp-flash.bin ${SECURITY_FW_VERSION}
+	hl_sec_fw ${BINS_DIR}/${DEV_PREFIX}-spi-otp-flash.bin ${SECURITY_FW_VERSION}
 
-	hl_sec_key ${DEV_PREFIX}-spi-otp-flash.bin hl 0
+	hl_sec_key ${BINS_DIR}/${DEV_PREFIX}-spi-otp-flash.bin hl 0
 
 	# generate 4 customer keys and hash them to spi security
 	for i in $(seq 1 $NUM_PROD_KEYS); do
 		CUST=cust${i}
-		hl_sec_key ${DEV_PREFIX}-spi-otp-flash.bin ${CUST} ${i}
+		hl_sec_key ${BINS_DIR}/${DEV_PREFIX}-spi-otp-flash.bin ${CUST} ${i}
 	done
 
 	# set revoke status as a 40bit (8bit * 5) binary value
-	hl_set_revoke_status ${DEV_PREFIX}-spi-otp-flash.bin
+	hl_set_revoke_status ${BINS_DIR}/${DEV_PREFIX}-spi-otp-flash.bin
 
-	cat ${DEV_PREFIX}-spi-otp-flash.bin	> ${DEV_PREFIX}-security-arc.bin.le
+	cat ${BINS_DIR}/${DEV_PREFIX}-spi-otp-flash.bin	> ${BINS_DIR}/${DEV_PREFIX}-security-arc.bin.le
 
-	swap_file ${DEV_PREFIX}-security-arc.bin.le ${DEV_PREFIX}-security-arc.bin.be
-	swap_file ${DEV_PREFIX}-erom.bin.le ${DEV_PREFIX}-erom.bin.be
-	swap_file ${DEV_PREFIX}-preboot.bin.le ${DEV_PREFIX}-preboot.bin.be
-	swap_file ${DEV_PREFIX}-ppboot-pcie_phy_fw.bin.le ${DEV_PREFIX}-ppboot-pcie_phy_fw.bin.be
-	swap_file  ${DEV_PREFIX}-images-pointers.bin.le  ${DEV_PREFIX}-images-pointers.bin.be
-	swap_file ${DEV_PREFIX}-fake-fuse.bin.le ${DEV_PREFIX}-fake-fuse.bin.be
-	swap_file ${DEV_PREFIX}-preboot_d1.bin.le ${DEV_PREFIX}-preboot_d1.bin.be
-	swap_file ${DEV_PREFIX}-sram-mbist-data.bin.le ${DEV_PREFIX}-sram-mbist-data.bin.be
+	swap_file ${BINS_DIR}/${DEV_PREFIX}-security-arc.bin.le ${BINS_DIR}/${DEV_PREFIX}-security-arc.bin.be
+	swap_file ${BINS_DIR}/${DEV_PREFIX}-erom.bin.le ${BINS_DIR}/${DEV_PREFIX}-erom.bin.be
+	swap_file ${BINS_DIR}/${DEV_PREFIX}-preboot.bin.le ${BINS_DIR}/${DEV_PREFIX}-preboot.bin.be
+	swap_file ${BINS_DIR}/${DEV_PREFIX}-ppboot-pcie_phy_fw.bin.le ${BINS_DIR}/${DEV_PREFIX}-ppboot-pcie_phy_fw.bin.be
+	swap_file ${BINS_DIR}/${DEV_PREFIX}-images-pointers.bin.le ${BINS_DIR}/${DEV_PREFIX}-images-pointers.bin.be
+	swap_file ${BINS_DIR}/${DEV_PREFIX}-fake-fuse.bin.le ${BINS_DIR}/${DEV_PREFIX}-fake-fuse.bin.be
+	swap_file ${BINS_DIR}/${DEV_PREFIX}-preboot_d1.bin.le ${BINS_DIR}/${DEV_PREFIX}-preboot_d1.bin.be
+	swap_file ${BINS_DIR}/${DEV_PREFIX}-sram-mbist-data.bin.le ${BINS_DIR}/${DEV_PREFIX}-sram-mbist-data.bin.be
 
-	calc_image_ptrs_sha384 ${DEV_PREFIX}-images-pointers.bin.be
+	calc_image_ptrs_sha384 ${BINS_DIR}/${DEV_PREFIX}-images-pointers.bin.be
 
-	rm -rf ${PPBOOT_PCIE_PHY_BINFILE}
+	rm -rf ${BINS_DIR}/${PPBOOT_PCIE_PHY_BINFILE}
 
 	## Convert bin file into Palladium hex format ##
 	# for otp flash
-	hl_pldm_flash_hex spi-otp-flash ${DEV_PREFIX}-spi-otp-flash.bin
+	hl_pldm_flash_hex spi-otp-flash ${BINS_DIR}/${DEV_PREFIX}-spi-otp-flash.bin ${HEXS_DIR}
 
 	# for preboot binary
-	swap_file ${PREBOOT_BINFILE} ${PREBOOT_BINFILE}.orig.be
-	hl_pldm_hex preboot ${PREBOOT_BINFILE}.orig.be 1
+	swap_file ${BINS_DIR}/${PREBOOT_BINFILE} ${BINS_DIR}/${PREBOOT_BINFILE}.orig.be
+	hl_pldm_hex preboot ${BINS_DIR}/${PREBOOT_BINFILE}.orig.be 1 ${HEXS_DIR}
 
-	PREBOOT_BIN_SZ=$(stat ${PREBOOT_BINFILE}.orig.be --format=%s)
+	PREBOOT_BIN_SZ=$(stat ${BINS_DIR}/${PREBOOT_BINFILE}.orig.be --format=%s)
 	PREBOOT_CHUNK_SZ=131072 #128kb
 	PREBOOT_CHUNKS=$((PREBOOT_BIN_SZ / PREBOOT_CHUNK_SZ + 1))
-	split -b$PREBOOT_CHUNK_SZ -d --additional-suffix=${PREBOOT_BINFILE}.split.be ${PREBOOT_BINFILE}.orig.be
+	split -b$PREBOOT_CHUNK_SZ -d --additional-suffix=${PREBOOT_BINFILE}.split.be ${BINS_DIR}${PREBOOT_BINFILE}.orig.be
 	for (( i=0; i<$PREBOOT_CHUNKS; i++ )); do
 		if test -f "x0${i}${PREBOOT_BINFILE}.split.be"; then
-			hl_pldm_hex preboot${i} x0${i}${PREBOOT_BINFILE}.split.be 1
+			hl_pldm_hex preboot${i} x0${i}${PREBOOT_BINFILE}.split.be 1 ${HEXS_DIR}
 		fi
 	done
 	# for preboot_d1 binary
@@ -521,6 +519,6 @@
 
 	# for ROM binary
 	hl_pad_arcboot_rom
-	hl_pldm_hex arcboot_rom ${ARCBOOT_BINFILE} 16
+	hl_pldm_hex arcboot_rom ${BINS_DIR}/${ARCBOOT_BINFILE} 16 ${HEXS_DIR}
 
 popd
diff --git a/board/habanalabs/common/post-image-spi.sh b/board/habanalabs/common/post-image-spi.sh
index a4568ab..ab5d4c1 100755
--- a/board/habanalabs/common/post-image-spi.sh
+++ b/board/habanalabs/common/post-image-spi.sh
@@ -78,7 +78,9 @@
 	printf "%08x" $img_loadl | rev_bytes_stdin
 
 	# Prepare checksum
-	local preboot_file=`echo ${PREBOOT_BINFILE} | cut -f1 -d"."`
+	local preboot_base=$(basename ${STAGE_DIR}/${PREBOOT_BINFILE})
+	local preboot_path=$(dirname ${STAGE_DIR}/${PREBOOT_BINFILE})
+	local preboot_file=${preboot_path}/${preboot_base%.*}
 	local preboot_le="${preboot_file}.bin.le"
 	local PRE_CKSUM=`${SHA_CHKSUM} ${preboot_le} | cut -d' ' -f1`
 	hl_chksum_add ${PRE_CKSUM}
@@ -90,8 +92,8 @@
 	local img_desc_sz=44
 
 	# Prepare the NEXT and MODIFIABLE for backward compatibility
-	echo "11110000" > ${HL_DEV}-next-image.txt
-	dd status=none conv=sync bs=8 count=1 if=${HL_DEV}-next-image.txt of=$img_file
+	echo "11110000" > ${STAGE_DIR}/${HL_DEV}-next-image.txt
+	dd status=none conv=sync bs=8 count=1 if=${STAGE_DIR}/${HL_DEV}-next-image.txt of=$img_file
 
 	# Pointers structure magic "NXT"0x0
 	printf "%x%x%x00" "'N" "'X" "'T" | rev_bytes_stdin | xxd -r -p -s +8 - $img_file
@@ -118,41 +120,40 @@
 pushd ${STAGE_DIR}
 	${HOST_DIR}/usr/bin/mkimage -C none -A arm64 -T script -d ${UBOOT_CMDFILE} boot.scr
 
-	handle_btl sectors-be.bin ${HL_DEV}-sectors.bin ${STAGE_DIR}
-	BTL_BINFILE=${HL_DEV}-sectors.bin
+	handle_btl ${STAGE_DIR}/sectors-be.bin ${BINS_DIR}/${HL_DEV}-sectors.bin ${STAGE_DIR}
+	BTL_BINFILE=${BINS_DIR}/${HL_DEV}-sectors.bin
 
-	dd status=none conv=sync bs=8 if=${PREBOOT_BINFILE}				 > ${HL_DEV}-preboot.bin.le
+	dd status=none conv=sync bs=8 if=${STAGE_DIR}/${PREBOOT_BINFILE} > ${BINS_DIR}/${HL_DEV}-preboot.bin.le
 	# Swap Preboot
-	swap_file ${HL_DEV}-preboot.bin.le ${HL_DEV}-preboot.bin.be
-	PREBOOT_BINFILE=${HL_DEV}-preboot.bin.be
-	hl_compress_bin ${HL_DEV}-preboot.bin.le
-
-	dd status=none conv=sync bs=8 if=u-boot.bin					 > ${HL_DEV}-u-boot.bin.le
+	swap_file ${BINS_DIR}/${HL_DEV}-preboot.bin.le ${BINS_DIR}/${HL_DEV}-preboot.bin.be
+	PREBOOT_BINFILE=${BINS_DIR}/${HL_DEV}-preboot.bin.be
+	hl_compress_bin ${BINS_DIR}${HL_DEV}-preboot.bin.le
+	dd status=none conv=sync bs=8 if=${IMGS_DIR}/u-boot.bin > ${BINS_DIR}/${HL_DEV}-u-boot.bin.le
 	# Swap U-boot
-	swap_file ${HL_DEV}-u-boot.bin.le ${HL_DEV}-u-boot.bin.be
-	UBOOT_BINFILE=${HL_DEV}-u-boot.bin.be
+	swap_file ${BINS_DIR}/${HL_DEV}-u-boot.bin.le ${BINS_DIR}/${HL_DEV}-u-boot.bin.be
+	UBOOT_BINFILE=${BINS_DIR}/${HL_DEV}-u-boot.bin.be
 
 	# Used by TRACE32 (Lauterbach JTAG)
-	dd status=none conv=sync bs=4 if=u-boot.bin					 > ${HL_DEV}-u-boot.bin
+	dd status=none conv=sync bs=4 if=${IMGS_DIR}/u-boot.bin > ${BINS_DIR}/${HL_DEV}-u-boot.bin
 
 	# Reserved space filled by 0xFF
-	dd status=none conv=sync bs=64K count=1 if=/dev/zero |  tr "\000" "\377"	 > 64K-0xff.bin
-	dd status=none conv=sync bs=128K count=1 if=/dev/zero | tr "\000" "\377"	 > 128K-0xff.bin
+	dd status=none conv=sync bs=64K count=1 if=/dev/zero |  tr "\000" "\377"	 > ${STAGE_DIR}/64K-0xff.bin
+	dd status=none conv=sync bs=128K count=1 if=/dev/zero | tr "\000" "\377"	 > ${STAGE_DIR}/128K-0xff.bin
 
 	# Metadata: Pointers
-	prep_image_ptrs ${HL_DEV}-images-pointers.bin.le
-	swap_file ${HL_DEV}-images-pointers.bin.le ${HL_DEV}-images-pointers.bin.be
-	dd status=none conv=sync bs=64K count=1 if=64K-0xff.bin				 > ${HL_DEV}-images-pointers.bin.be.spi
-	dd status=none conv=sync,notrunc bs=4 if=${HL_DEV}-images-pointers.bin.be of=${HL_DEV}-images-pointers.bin.be.spi
+	prep_image_ptrs ${BINS_DIR}$/{HL_DEV}-images-pointers.bin.le
+	swap_file ${BINS_DIR}/${HL_DEV}-images-pointers.bin.le ${BINS_DIR}/${HL_DEV}-images-pointers.bin.be
+	dd status=none conv=sync bs=64K count=1 if=${STAGE_DIR}/64K-0xff.bin > ${STAGE_DIR}/${HL_DEV}-images-pointers.bin.be.spi
+	dd status=none conv=sync,notrunc bs=4 if=${BINS_DIR}/${HL_DEV}-images-pointers.bin.be of=${STAGE_DIR}${HL_DEV}-images-pointers.bin.be.spi
 
-	create_spi_image ${BTL_BINFILE} ${PREBOOT_BINFILE} ${HL_DEV}-images-pointers.bin.be.spi ${HL_DEV}-spi-flash.bin.nocrc
+	create_spi_image ${BTL_BINFILE} ${STAGE_DIR}/${PREBOOT_BINFILE} ${STAGE_DIR}/${HL_DEV}-images-pointers.bin.be.spi ${STAGE_DIR}/${HL_DEV}-spi-flash.bin.nocrc
 
 	# Prepare checksum
-	CKSUM=`${SHA_CHKSUM} ${HL_DEV}-spi-flash.bin.nocrc | cut -d' ' -f1`
-	echo ${CKSUM} | xxd -p -r							 > ${HL_DEV}-spi-flash.crc
-	dd status=none conv=sync bs=64K		if=${HL_DEV}-spi-flash.bin.nocrc	 > ${HL_DEV}-spi-flash.bin
-	dd status=none conv=sync bs=32		if=${HL_DEV}-spi-flash.crc		>> ${HL_DEV}-spi-flash.bin
+	CKSUM=`${SHA_CHKSUM} ${STAGE_DIR}/${HL_DEV}-spi-flash.bin.nocrc | cut -d' ' -f1`
+	echo ${CKSUM} | xxd -p -r									> ${STAGE_DIR}/${HL_DEV}-spi-flash.crc
+	dd status=none conv=sync bs=64K		if=${STAGE_DIR}/${HL_DEV}-spi-flash.bin.nocrc		> ${BINS_DIR}/${HL_DEV}-spi-flash.bin
+	dd status=none conv=sync bs=32		if=${STAGE_DIR}/${HL_DEV}-spi-flash.crc			>> ${BINS_DIR}/${HL_DEV}-spi-flash.bin
 
 	# Convert bin file into Palladium hex format
-	xxd -p -c1 -g1				${HL_DEV}-spi-flash.bin	 		 > ${HL_DEV}-spi-flash.hex
+	xxd -p -c1 -g1				${BINS_DIR}/${HL_DEV}-spi-flash.bin	 		 > ${HEXS_DIR}/${HL_DEV}-spi-flash.hex
 popd
diff --git a/board/habanalabs/fs1_arc/fs1-post-image-fit-arc.sh b/board/habanalabs/fs1_arc/fs1-post-image-fit-arc.sh
index c773471..0e1a122 100755
--- a/board/habanalabs/fs1_arc/fs1-post-image-fit-arc.sh
+++ b/board/habanalabs/fs1_arc/fs1-post-image-fit-arc.sh
@@ -13,7 +13,7 @@
 HL_VERSION=`cut -d- -f4 ${BR2_EXTERNAL}/output/${HL_DEV}/build/zephyr-custom/localversion-habanalabs`
 
 ZEPHYR_ITS_FILE=zephyr-${DEV_NAME}-${HL_TARGET}.its
-ZEPHYR_ITB_FILE=zephyr-fit.itb
+ZEPHYR_ITB_FILE=${ITBS_DIR}/zephyr-fit.itb
 
 # Set compression to lzo
 HL_COMPRESS=lzo
@@ -61,5 +61,5 @@
 	sed ${ITS_SED_PARAMS} ${ZEPHYR_ITS_TEMPLATE} > ${ZEPHYR_ITS_FILE}
 	${MKIMAGE} -D "-W no-unit_address_vs_reg" -f ${ZEPHYR_ITS_FILE} -k ${STAGE_DIR} ${ZEPHYR_ITB_FILE}
 
-	dd status=none conv=sync bs=4 if=${ZEPHYR_ITB_FILE} > ${DEV_PREFIX}-boot-fit.itb
+	dd status=none conv=sync bs=4 if=${ZEPHYR_ITB_FILE} > ${ITBS}/${DEV_PREFIX}-boot-fit.itb
 popd
diff --git a/board/habanalabs/fs1_arc/fs1-post-image-spi-arc_global.sh b/board/habanalabs/fs1_arc/fs1-post-image-spi-arc_global.sh
index b45df1a..31f32be 100755
--- a/board/habanalabs/fs1_arc/fs1-post-image-spi-arc_global.sh
+++ b/board/habanalabs/fs1_arc/fs1-post-image-spi-arc_global.sh
@@ -58,7 +58,6 @@
 SECURITY_VER_FILE=${SPECS_EMB_PROJ_DIR}/${DEV_PREFIX}/${DEV_PREFIX}_security.h
 export BOOT_POINTER_HEADER=${SPECS_EMB_PROJ_DIR}/common/hl_boot_desc.h
 
-source ${DEVICE_BASE}/fs1-image-utility.sh
 source ${DEVICE_BASE}/post-image-spi.sh
 
 SECURITY_FW_VERSION=$(grep -owE "HL_SEC_FW_VERSION[[:space:]]+[0-9]+" $SECURITY_VER_FILE | sed 's/^.*[ \t]\([0-9]\+\)/\1/;')
@@ -115,16 +114,16 @@
         KEY_OFFSET=${3}
         KEY_BASE_NAME_PUB=${KEY_BASE_NAME}.pub
 
-	calculate_hash ${NAME}_${KEY_BASE_NAME_PUB} ${NAME}_${KEY_BASE_NAME_PUB}.hash
-	cat ${NAME}_${KEY_BASE_NAME_PUB}.hash				> ${FILE}.tmp
+	calculate_hash ${NAME}_${KEY_BASE_NAME_PUB} ${STAGE_DIR}/${NAME}_${KEY_BASE_NAME_PUB}.hash
+	cat ${STAGE_DIR}/${NAME}_${KEY_BASE_NAME_PUB}.hash				> ${FILE}.tmp
 	dd status=none conv=notrunc seek=$(( ${KEY_OFFSET}*2 + 1 )) bs=16 if=${FILE}.tmp of=${FILE}
 
 	if [ ${HL_TARGET} == "pldm" ]; then
 		for i in $(seq 1 $HASH_SZ); do
-			DWORD=$(xxd -e -c32 -g4 ${NAME}_${KEY_BASE_NAME_PUB}.hash | cut -d' ' -f$(( ${i} + 2 )))
+			DWORD=$(xxd -e -c32 -g4 ${STAGE_DIR}/${NAME}_${KEY_BASE_NAME_PUB}.hash | cut -d' ' -f$(( ${i} + 2 )))
 			OFFSET=$(( $(( ${KEY_OFFSET}*8 )) +  ${i} ))
 			FORCE="{h5_core.h5_psoc_top.psoc_macro.efuse.efuse_cfg_reg.u_efuse_cfg_reg_fields.u_cfg_reg_base_data_$(( 18 + ${OFFSET}))_r.data_in_r[31:0]}"
-			echo "force ${FORCE} 'h${DWORD}" >> ${HL_TARGET}-force.hex
+			echo "force ${FORCE} 'h${DWORD}" >> ${STAGE_DIR}/${HL_TARGET}-force.hex
 		done
 	fi
 }
@@ -132,9 +131,9 @@
 function hl_key_revoke {
 	REVOKE_STATUS=$1
 	if [ $REVOKE_STATUS -eq 1 ]; then
-		echo -n fe >> revoke_status.hex
+		echo -n fe >> ${STAGE_DIR}/revoke_status.hex
 	else
-		echo -n ff >> revoke_status.hex
+		echo -n ff >> ${STAGE_DIR}/revoke_status.hex
 	fi
 }
 
@@ -152,7 +151,7 @@
 	DWORD=$(( DWORD + $((${FW_VERSION_LOCATION}<<13))))
 	DWORD=$(( DWORD + $((${SPI_CONTROLLER}<<17))))
 	DWORD=$(printf "%08x" ${DWORD})
-	echo "force {h5_core.h5_psoc_top.psoc_macro.efuse.efuse_cfg_reg.u_efuse_cfg_reg_fields.u_cfg_reg_base_data_58_r.data_in_r[31:0]} 'h${DWORD}" > ${HL_TARGET}-force.hex
+	echo "force {h5_core.h5_psoc_top.psoc_macro.efuse.efuse_cfg_reg.u_efuse_cfg_reg_fields.u_cfg_reg_base_data_58_r.data_in_r[31:0]} 'h${DWORD}" > ${HEXS_DIR}${HL_TARGET}-force.hex
 }
 
 pushd ${STAGE_DIR}
@@ -168,7 +167,7 @@
 	done
 
 	## create otp spi image
-	dd status=none conv=sync count=12 bs=16 if=/dev/zero | tr '\000' $'\xFF' > ${DEV_PREFIX}-spi-otp-flash.bin
+	dd status=none conv=sync count=12 bs=16 if=/dev/zero | tr '\000' $'\xFF' > ${BINS_DIR}/${DEV_PREFIX}-spi-otp-flash.bin
 
 	# get private and public keys
 	copy_all_keys
@@ -177,13 +176,13 @@
 	generate_public_keys
 
 	# Create erom binaries
-	prep_image_erom ${DEV_PREFIX}-erom.bin.le
+	prep_image_erom ${BINS_DIR}/${DEV_PREFIX}-erom.bin.le
 
 	# Create fake fuse binary
-	prep_image_fake_fuse ${DEV_PREFIX}-fake-fuse.bin.le
+	prep_image_fake_fuse ${BINS_DIR}/${DEV_PREFIX}-fake-fuse.bin.le
 
 	# Pad preboot binary and then pack
-	zephyr_preboot_pack ${PREBOOT_BINFILE} ${PRIME_IMAGE_RAM_KEY} ${PRIME_IMAGE_RAM_CRC_ERROR} ${PRIME_IMAGE_RAM_SIGN_ERROR}
+	zephyr_preboot_pack ${BINS_DIR}/${PREBOOT_BINFILE} ${PRIME_IMAGE_RAM_KEY} ${PRIME_IMAGE_RAM_CRC_ERROR} ${PRIME_IMAGE_RAM_SIGN_ERROR}
 
 	# create combined name of ppboot and pcie
 	PPBOOT=$(echo "$PPBOOT_BINFILE" | cut -f 1 -d '.')
@@ -191,47 +190,46 @@
 	PPBOOT_PCIE_PHY_BINFILE=${PPBOOT}-${PCIE_PHY}.bin
 
 	# concatenate both FW
-	hl_join_bin ${PPBOOT} ${PCIE_PHY} ${PPBOOT_PCIE_PHY_BINFILE}
+	hl_join_bin ${BINS_DIR}/${PPBOOT} ${PCIE_PHY} ${BINS_DIR}/${PPBOOT_PCIE_PHY_BINFILE}
 
 	# pack and sign image
-	hl_pack_bin ${PPBOOT_PCIE_PHY_BINFILE} 64K ${PRIME_IMAGE_PPBOOT_KEY} ${PRIME_IMAGE_PPBOOT_CRC_ERROR} ${PRIME_IMAGE_PPBOOT_SIGN_ERROR}
+	hl_pack_bin ${BINS_DIR}/${PPBOOT_PCIE_PHY_BINFILE} 64K ${PRIME_IMAGE_PPBOOT_KEY} ${PRIME_IMAGE_PPBOOT_CRC_ERROR} ${PRIME_IMAGE_PPBOOT_SIGN_ERROR}
 
 	# Create image pointers
-	prep_image_ptrs ${DEV_PREFIX}-images-pointers.bin.le
+	prep_image_ptrs ${BINS_DIR}/${DEV_PREFIX}-images-pointers.bin.le
 
 	if [ ${HL_TARGET} == "pldm" ]; then
 		hl_efuse_sys_config_generate
 	fi
+	swap_file ${BINS_DIR}/${DEV_PREFIX}-ppboot-pcie_phy_fw.bin.le ${BINS_DIR}/${DEV_PREFIX}-ppboot-pcie_phy_fw.bin.be
+	swap_file ${BINS_DIR}/${DEV_PREFIX}-images-pointers.bin.le  ${BINS_DIR}/${DEV_PREFIX}-images-pointers.bin.be
+	swap_file ${BINS_DIR}/${DEV_PREFIX}-fake-fuse.bin.le ${BINS_DIR}/${DEV_PREFIX}-fake-fuse.bin.be
+	swap_file ${BINS_DIR}/${DEV_PREFIX}-preboot.bin.le ${BINS_DIR}/${DEV_PREFIX}-preboot.bin.be
 
-	swap_file ${DEV_PREFIX}-ppboot-pcie_phy_fw.bin.le ${DEV_PREFIX}-ppboot-pcie_phy_fw.bin.be
-	swap_file ${DEV_PREFIX}-images-pointers.bin.le  ${DEV_PREFIX}-images-pointers.bin.be
-	swap_file ${DEV_PREFIX}-fake-fuse.bin.le ${DEV_PREFIX}-fake-fuse.bin.be
-	swap_file ${DEV_PREFIX}-preboot.bin.le ${DEV_PREFIX}-preboot.bin.be
-
-	create_spi_image ${DEV_PREFIX}-spi-flash.bin.nocrc \
-			 ${DEV_PREFIX}-ppboot-pcie_phy_fw.bin.${ENDIANNESS} ${DEV_PREFIX}-preboot.bin.${ENDIANNESS} \
-			 ${DEV_PREFIX}-images-pointers.bin.${ENDIANNESS} \
-			 ${DEV_PREFIX}-fake-fuse.bin.${ENDIANNESS}
+	create_spi_image ${BINS_DIR}/${DEV_PREFIX}-spi-flash.bin.nocrc \
+			 ${BINS_DIR}/${DEV_PREFIX}-ppboot-pcie_phy_fw.bin.${ENDIANNESS} \
+			 ${BINS_DIR}/${DEV_PREFIX}-preboot.bin.${ENDIANNESS} \
+			 ${BINS_DIR}/${DEV_PREFIX}-images-pointers.bin.${ENDIANNESS} \
+			 ${BINS_DIR}/${DEV_PREFIX}-fake-fuse.bin.${ENDIANNESS}
 
 	# SPI flash binary is provided as Big Endian, for manufacturing programmer.
-	swap_file ${DEV_PREFIX}-spi-flash.bin.nocrc ${DEV_PREFIX}-spi-flash.bin.nocrc.le
+	swap_file ${BINS_DIR}/${DEV_PREFIX}-spi-flash.bin.nocrc ${BINS_DIR}/${DEV_PREFIX}-spi-flash.bin.nocrc.le
 
 	# Prepare checksum
-	CKSUM=`${SHA256_CHKSUM} ${DEV_PREFIX}-spi-flash.bin.nocrc.le | cut -d' ' -f1`
-	echo ${CKSUM} | xxd -p -r								> ${DEV_PREFIX}-spi-flash.crc
-	dd status=none conv=sync bs=64K			if=${DEV_PREFIX}-spi-flash.bin.nocrc.le	> ${DEV_PREFIX}-spi-flash.bin
-	dd status=none conv=sync bs=$SHA256_SIZE	if=${DEV_PREFIX}-spi-flash.crc		>> ${DEV_PREFIX}-spi-flash.bin
+	CKSUM=`${SHA256_CHKSUM} ${BINS_DIR}/${DEV_PREFIX}-spi-flash.bin.nocrc.le | cut -d' ' -f1`
+	echo ${CKSUM} | xxd -p -r								> ${BINS_DIR}/${DEV_PREFIX}-spi-flash.crc
+	dd status=none conv=sync bs=64K			if=${BINS_DIR}/${DEV_PREFIX}-spi-flash.bin.nocrc.le	> ${BINS_DIR}/${DEV_PREFIX}-spi-flash.bin
+	dd status=none conv=sync bs=$SHA256_SIZE	if=${BINS_DIR}/${DEV_PREFIX}-spi-flash.crc		>> ${BINS_DIR}/${DEV_PREFIX}-spi-flash.bin
 
-	rm -rf ${PPBOOT_PCIE_PHY_BINFILE}
+	rm -rf ${BINS_DIR}/${PPBOOT_PCIE_PHY_BINFILE}
 
 	## Convert bin file into Palladium hex format ##
 	# for SPI flash
-	hl_pldm_flash_hex spi-flash ${DEV_PREFIX}-spi-flash.bin.nocrc
+	hl_pldm_flash_hex spi-flash ${BINS_DIR}/${DEV_PREFIX}-spi-flash.bin.nocrc ${HEXS_DIR}
 
 	# for preboot binary
-	swap_file ${PREBOOT_BINFILE} ${PREBOOT_BINFILE}.orig.be
-	hl_pldm_hex preboot ${PREBOOT_BINFILE}.orig.be 1
-
+	swap_file ${BINS_DIR}/${PREBOOT_BINFILE} ${BINS_DIR}/${PREBOOT_BINFILE}.orig.be
+	hl_pldm_hex preboot ${BINS_DIR}/${PREBOOT_BINFILE}.orig.be 1 ${HEXS_DIR}
 	# for ROM binary
 	hl_pad_arcboot_rom
 popd
diff --git a/board/habanalabs/fs1_arc/post-image-spi.sh b/board/habanalabs/fs1_arc/post-image-spi.sh
index cbd3cde..153cff3 100755
--- a/board/habanalabs/fs1_arc/post-image-spi.sh
+++ b/board/habanalabs/fs1_arc/post-image-spi.sh
@@ -194,14 +194,16 @@
 MAX_SPI_DMA_DATA_TRANSFER=65520
 SPI_DMA_DEST_64KB_OFFS=65536
 function hl_pack_bin {
-	FILEPATH=$1
+	FILEFULLPATH=$1
+	FILENAME=$(basename ${FILEFULLPATH})
+	FILEPATH=$(dirname ${FILEFULLPATH})
 	BSIZE=$2
 	KEY_NUM=$3
 	CRC_ERROR=$4
 	SIGN_ERROR=$5
 
-	PADFULLBINFILE=${FILEPATH}.pad
-	dd status=none conv=sync bs=16 if=${FILEPATH} of=${PADFULLBINFILE}
+	PADFULLBINFILE=${FILEFULLPATH}.pad
+	dd status=none conv=sync bs=16 if=${FILEFULLPATH} of=${PADFULLBINFILE}
 
 	SIZE_WITHOUT_CRC=$(stat --printf="%s" ${PADFULLBINFILE})
 	NUM_SLICES=$((SIZE_WITHOUT_CRC/MAX_SPI_DMA_DATA_TRANSFER))
@@ -252,13 +254,13 @@
 		fi
 		flash_offset=$((SPI_DMA_DEST_64KB_OFFS/4))
 		flash_offset=$((flash_offset*slice_cnt))
-		dd status=none conv=sync if=${BIN_FILE} bs=4 seek=$flash_offset of=${DEV_PREFIX}-${FILEPATH}.le.int
+		dd status=none conv=sync if=${BIN_FILE} bs=4 seek=$flash_offset of=${STAGE_DIR}/${DEV_PREFIX}-${FILENAME}.le.int
 		leftover=$((leftover-size_bytes))
 		skip_offset=$((skip_offset+size_words))
 	done
 
-	hl_reverse_words ${DEV_PREFIX}-${FILEPATH}.le.int ${DEV_PREFIX}-${FILEPATH}.le 4
-	rm -f ${DEV_PREFIX}-${FILEPATH}.le.int
+	hl_reverse_words ${STAGE_DIR}/${DEV_PREFIX}-${FILENAME}.le.int ${FILEPATH}/${DEV_PREFIX}-${FILENAME}.le 4
+	rm -f ${STAGE_DIR}/${DEV_PREFIX}-${FILENAME}.le.int
 }
 
 function hl_sign_bin {
@@ -272,7 +274,7 @@
 	PROP_NAME=$2
 	OUT_NAME=$3
 
-	TMPNAME=${KEY_NAME}-${PROP_NAME}.txt
+	TMPNAME=${STAGE_DIR}/${KEY_NAME}-${PROP_NAME}.txt
 	fdtget ${KEY_NAME}.dtb /signature/key-${KEY_NAME} rsa,${PROP_NAME} -t bx > ${TMPNAME}
 	sed 's/\s\+/\n/g' ${TMPNAME} > ${TMPNAME}.tmp
 	sed -e 's/\b[0-9]\b/0&/' -e 's/\b[a-f]\b/0&/' ${TMPNAME}.tmp  > ${TMPNAME}.tmp.le
@@ -289,24 +291,24 @@
 		KEY_NAME=$(echo -n "cust${KEY_NUM}_${KEY_BASE_NAME}")
 	fi
 
-	extract_key_props ${KEY_NAME} r-squared rr.prop
+	extract_key_props ${KEY_NAME} r-squared ${STAGE_DIR}/rr.prop
 
-	extract_key_props ${KEY_NAME} modulus mod.prop
+	extract_key_props ${KEY_NAME} modulus ${STAGE_DIR}/mod.prop
 
-	extract_key_props ${KEY_NAME} exponent exp.prop
+	extract_key_props ${KEY_NAME} exponent ${STAGE_DIR}/exp.prop
 
-	extract_key_props ${KEY_NAME} n0-inverse inv.proptemp
-	swap_file inv.proptemp inv.prop
+	extract_key_props ${KEY_NAME} n0-inverse ${STAGE_DIR}/inv.proptemp
+	swap_file ${STAGE_DIR}/inv.proptemp ${STAGE_DIR}/inv.prop
 
-	extract_key_props ${KEY_NAME} num-bits nb.proptemp
-	swap_file nb.proptemp nb.prop
+	extract_key_props ${KEY_NAME} num-bits ${STAGE_DIR}/nb.proptemp
+	swap_file ${STAGE_DIR}/nb.proptemp ${STAGE_DIR}/nb.prop
 
 	#no separate public key will be part of the image, since these props serve the purpose.
 	# Hash will be computed on the Props itself
-	cat rr.prop mod.prop exp.prop inv.prop nb.prop	> ${KEY_NAME}.pub
-	cat rr.prop mod.prop exp.prop inv.prop nb.prop	> ${KEY_NAME}-key-props.bin
+	cat ${STAGE_DIR}/rr.prop ${STAGE_DIR}/mod.prop ${STAGE_DIR}/exp.prop ${STAGE_DIR}/inv.prop ${STAGE_DIR}/nb.prop	> ${STAGE_DIR}/${KEY_NAME}.pub
+	cat ${STAGE_DIR}/rr.prop ${STAGE_DIR}/mod.prop ${STAGE_DIR}/exp.prop ${STAGE_DIR}/inv.prop ${STAGE_DIR}/nb.prop	> ${STAGE_DIR}/${KEY_NAME}-key-props.bin
 
-	rm -f *.txt *.proptemp *.prop
+	rm -f ${STAGE_DIR}/*.txt ${STAGE_DIR}/*.proptemp ${STAGE_DIR}/*.prop
 }
 
 function hl_chksum_padding {
diff --git a/board/habanalabs/gaudi2/post-image-common.sh b/board/habanalabs/gaudi2/post-image-common.sh
index 3a94a6a..c21e076 100755
--- a/board/habanalabs/gaudi2/post-image-common.sh
+++ b/board/habanalabs/gaudi2/post-image-common.sh
@@ -13,7 +13,7 @@
 	PROP_NAME=$2
 	OUT_NAME=$3
 
-	TMPNAME=${KEY_NAME}-${PROP_NAME}.txt
+	TMPNAME=${STAGE_DIR}/${KEY_NAME}-${PROP_NAME}.txt
 	fdtget ${KEY_NAME}.dtb /signature/key-${KEY_BASE_NAME} rsa,${PROP_NAME} -t bx > ${TMPNAME}
 	sed 's/\s\+/\n/g' ${TMPNAME} > ${TMPNAME}.tmp
 	sed -e 's/\b[0-9]\b/0&/' -e 's/\b[a-f]\b/0&/' ${TMPNAME}.tmp  > ${TMPNAME}.tmp.le
@@ -25,21 +25,21 @@
 	KEY_NAME=$1
 	PREFIX=$2
 
-	hl_sec_extract_key_props ${KEY_NAME} r-squared ${PREFIX}_rr.prop
+	hl_sec_extract_key_props ${KEY_NAME} r-squared ${STAGE_DIR}/${PREFIX}_rr.prop
 
-	hl_sec_extract_key_props ${KEY_NAME} modulus ${PREFIX}_mod.prop
+	hl_sec_extract_key_props ${KEY_NAME} modulus ${STAGE_DIR}/${PREFIX}_mod.prop
 
-	hl_sec_extract_key_props ${KEY_NAME} exponent ${PREFIX}_exp.prop
+	hl_sec_extract_key_props ${KEY_NAME} exponent ${STAGE_DIR}/${PREFIX}_exp.prop
 
-	hl_sec_extract_key_props ${KEY_NAME} n0-inverse ${PREFIX}_inv.proptemp
-	swap_file ${PREFIX}_inv.proptemp ${PREFIX}_inv.prop
+	hl_sec_extract_key_props ${KEY_NAME} n0-inverse ${STAGE_DIR}/${PREFIX}_inv.proptemp
+	swap_file ${STAGE_DIR}/${PREFIX}_inv.proptemp ${STAGE_DIR}/${PREFIX}_inv.prop
 
-	hl_sec_extract_key_props ${KEY_NAME} num-bits ${PREFIX}_nb.proptemp
-	swap_file ${PREFIX}_nb.proptemp ${PREFIX}_nb.prop
+	hl_sec_extract_key_props ${KEY_NAME} num-bits ${STAGE_DIR}/${PREFIX}_nb.proptemp
+	swap_file ${STAGE_DIR}/${PREFIX}_nb.proptemp ${STAGE_DIR}/${PREFIX}_nb.prop
 
 	# no separate public key will be part of the image, since these props serve the purpose.
 	# Hash will be computed on the Props itself
-	cat ${PREFIX}_rr.prop ${PREFIX}_mod.prop ${PREFIX}_exp.prop ${PREFIX}_inv.prop ${PREFIX}_nb.prop > ${KEY_NAME}.pub
+	cat ${STAGE_DIR}/${PREFIX}_rr.prop ${STAGE_DIR}/${PREFIX}_mod.prop ${STAGE_DIR}/${PREFIX}_exp.prop ${STAGE_DIR}/${PREFIX}_inv.prop ${STAGE_DIR}/${PREFIX}_nb.prop > ${STAGE_DIR}/${KEY_NAME}.pub
 
-	rm -f ${PREFIX}_*.proptemp ${PREFIX}_*.prop
+	rm -f ${STAGE_DIR}/${PREFIX}_*.proptemp ${STAGE_DIR}/${PREFIX}_*.prop
 }
diff --git a/board/habanalabs/gaudi2/post-image-fit-descriptor.sh b/board/habanalabs/gaudi2/post-image-fit-descriptor.sh
index e68f2fa..db93a39 100644
--- a/board/habanalabs/gaudi2/post-image-fit-descriptor.sh
+++ b/board/habanalabs/gaudi2/post-image-fit-descriptor.sh
@@ -5,8 +5,8 @@
 
 function hl_handle_desc
 {
-	openssl dgst -sha384 -binary ${HL_DEV}-fit.itb > img-hash-${HL_DEV}-fit.sha384
-	openssl dgst -sha384 -binary ${HL_DEV}-boot-fit.itb > img-hash-${HL_DEV}-boot-fit.sha384
-	add_descriptor_to_fit ${HL_DEV}-fit.itb
-	add_descriptor_to_fit ${HL_DEV}-boot-fit.itb
+	openssl dgst -sha384 -binary ${ITBS_DIR}${HL_DEV}-fit.itb > ${ITBS_DIR}img-hash-${HL_DEV}-fit.sha384
+	openssl dgst -sha384 -binary ${ITBS_DIR}${HL_DEV}-boot-fit.itb > ${ITBS_DIR}img-hash-${HL_DEV}-boot-fit.sha384
+	add_descriptor_to_fit ${ITBS_DIR}${HL_DEV}-fit.itb
+	add_descriptor_to_fit ${ITBS_DIR}${HL_DEV}-boot-fit.itb
 }
diff --git a/board/habanalabs/gaudi2_arc/post-image-spi.sh b/board/habanalabs/gaudi2_arc/post-image-spi.sh
index 71f04f8..a6a4f45 100755
--- a/board/habanalabs/gaudi2_arc/post-image-spi.sh
+++ b/board/habanalabs/gaudi2_arc/post-image-spi.sh
@@ -43,7 +43,7 @@
 
 pushd ${STAGE_DIR}
 	# for SecureBoot ROM
-	hl_pldm_hex arcboot_rom arcboot_rom.bin 16
+	hl_pldm_hex arcboot_rom arcboot_rom.bin 16 ${HEXS_DIR}
 
 	hl_create_zephyr_bins_config
 	hl_extract_zephyr_elf_sections
@@ -53,8 +53,8 @@
 	# hl_copy_binaries function is called from post-image-spi-arc_global.sh
 	# common script. All the binaries required for creation of spi
 	# are compiled as part of gaudi2_arc HL_TARGET
-	touch ${DEV_PREFIX}-preboot_d1.bin.le
-	touch ${DEV_PREFIX}-sram-mbist-data.bin.le
+	touch ${BINS_DIR}/${DEV_PREFIX}-preboot_d1.bin.le
+	touch ${BINS_DIR}/${DEV_PREFIX}-sram-mbist-data.bin.le
 	return
 }
 
@@ -83,11 +83,11 @@
 	OUT_NAME=$3
 
 	TMPNAME=${KEY_NAME}-${PROP_NAME}.txt
-	fdtget ${KEY_NAME}.dtb /signature/key-${KEY_NAME} rsa,${PROP_NAME} -t bx > ${TMPNAME}
-	sed 's/\s\+/\n/g' ${TMPNAME} > ${TMPNAME}.tmp
-	sed -e 's/\b[0-9]\b/0&/' -e 's/\b[a-f]\b/0&/' ${TMPNAME}.tmp  > ${TMPNAME}.tmp.le
-	xxd -r -p ${TMPNAME}.tmp.le > ${OUT_NAME}
-	rm -f ${TMPNAME}.*
+	fdtget ${KEY_NAME}.dtb /signature/key-${KEY_NAME} rsa,${PROP_NAME} -t bx > ${STAGE_DIR}/${TMPNAME}
+	sed 's/\s\+/\n/g' ${STAGE_DIR}/${TMPNAME} > ${STAGE_DIR}/${TMPNAME}.tmp
+	sed -e 's/\b[0-9]\b/0&/' -e 's/\b[a-f]\b/0&/' ${STAGE_DIR}/${TMPNAME}.tmp > ${STAGE_DIR}/${TMPNAME}.tmp.le
+	xxd -r -p ${STAGE_DIR}/${TMPNAME}.tmp.le > ${OUT_NAME}
+	rm -f ${STAGE_DIR}/${TMPNAME}.*
 }
 
 function hl_get_key_props {
@@ -100,28 +100,28 @@
 
 	# second DWORD is dummy, used only for alignment
 	INDEX=$(printf "%08x%08x" ${KEY_NUM} ${KEY_NUM})
-	echo ${INDEX} | xxd -r -p > key_header.tmp
+	echo ${INDEX} | xxd -r -p > ${STAGE_DIR}/key_header.tmp
 	# key header is needed in little endian, so swap it
-	swap_file key_header.tmp key_header.prop
+	swap_file ${STAGE_DIR}/key_header.tmp ${STAGE_DIR}/key_header.prop
 
-	extract_key_props ${KEY_NAME} r-squared rr.prop
+	extract_key_props ${KEY_NAME} r-squared ${STAGE_DIR}/rr.prop
 
-	extract_key_props ${KEY_NAME} modulus mod.prop
+	extract_key_props ${KEY_NAME} modulus ${STAGE_DIR}/mod.prop
 
-	extract_key_props ${KEY_NAME} exponent exp.prop
+	extract_key_props ${KEY_NAME} exponent ${STAGE_DIR}/exp.prop
 
-	extract_key_props ${KEY_NAME} n0-inverse inv.proptemp
-	swap_file inv.proptemp inv.prop
+	extract_key_props ${KEY_NAME} n0-inverse ${STAGE_DIR}/inv.proptemp
+	swap_file ${STAGE_DIR}/inv.proptemp ${STAGE_DIR}/inv.prop
 
-	extract_key_props ${KEY_NAME} num-bits nb.proptemp
-	swap_file nb.proptemp nb.prop
+	extract_key_props ${KEY_NAME} num-bits ${STAGE_DIR}/nb.proptemp
+	swap_file ${STAGE_DIR}/nb.proptemp ${STAGE_DIR}/nb.prop
 
 	#no separate public key will be part of the image, since these props serve the purpose.
 	# Hash will be computed on the Props itself
-	cat rr.prop mod.prop exp.prop inv.prop nb.prop	> ${KEY_NAME}.pub
-	cat rr.prop mod.prop exp.prop inv.prop nb.prop key_header.prop > ${KEY_NAME}-key-props.bin
+	cat ${STAGE_DIR}/rr.prop ${STAGE_DIR}/mod.prop ${STAGE_DIR}/exp.prop ${STAGE_DIR}/inv.prop ${STAGE_DIR}/nb.prop	> ${STAGE_DIR}/${KEY_NAME}.pub
+	cat ${STAGE_DIR}/rr.prop ${STAGE_DIR}/mod.prop ${STAGE_DIR}/exp.prop ${STAGE_DIR}/inv.prop ${STAGE_DIR}/nb.prop ${STAGE_DIR}/key_header.prop > ${STAGE_DIR}/${KEY_NAME}-key-props.bin
 
-	rm -f *.txt *.proptemp *.prop
+	rm -f ${STAGE_DIR}/*.txt ${STAGE_DIR}/*.proptemp ${STAGE_DIR}/*.prop
 }
 
 # Generate the Public Key, whose hash is needed to be stored
@@ -133,7 +133,9 @@
 }
 
 function hl_pack_bin {
-	FILEPATH=$1
+	FILEFULLPATH=$1
+	FILENAME=$(basename ${FILEFULLPATH})
+	FILEPATH=$(dirname ${FILEFULLPATH})
 	BSIZE=$2
 	KEY_NUM=$3
 	CRC_ERROR=$4
@@ -147,32 +149,33 @@
 	#Extract Key Properties and store them in a bin file
 	hl_get_key_props ${KEY_NUM}
 
-	TEMPNAME=${FUNCNAME}-$(basename ${FILEPATH})
-	dd status=none conv=sync bs=16 if=${FILEPATH}			> ${TEMPNAME}
+	TEMPNAME=${FUNCNAME}-${FILENAME}
+	TEMPFULLPATH=${STAGE_DIR}/${FUNCNAME}-${FILENAME}
+	dd status=none conv=sync bs=16 if=${FILEFULLPATH} > ${TEMPFULLPATH}
 
-	append_sha384 ${TEMPNAME}
+	append_sha384 ${TEMPFULLPATH}
 
-	hl_get_ver ${TEMPNAME}
+	hl_get_ver ${TEMPFULLPATH}
 	VPKFILE=${TEMPNAME}.ver.payload.key
 
 	#No need to pack Public Key, since Key Props are enough
-	cat ${TEMPNAME}.ver ${TEMPNAME} ${KEY_NAME}-key-props.bin > ${VPKFILE}
+	cat ${TEMPFULLPATH}.ver ${TEMPFULLPATH} ${STAGE_DIR}/${KEY_NAME}-key-props.bin > ${STAGE_DIR}/${VPKFILE}
 
-	$OPENSSL dgst -sha384 -binary ${VPKFILE} > psk_${VPKFILE}.sha384
+	$OPENSSL dgst -sha384 -binary ${STAGE_DIR}/${VPKFILE} > ${STAGE_DIR}/psk_${VPKFILE}.sha384
 
-	hl_sign_bin ${VPKFILE} ${KEY_NAME}
+	hl_sign_bin ${STAGE_DIR}/${VPKFILE} ${STAGE_DIR}/${KEY_NAME}
 	if [ ${SIGN_ERROR} -eq 1 ]; then
-		dd status=none conv=sync count=24 bs=16 if=/dev/urandom	> ${VPKFILE}.sig
+		dd status=none conv=sync count=24 bs=16 if=/dev/urandom	> ${STAGE_DIR}/${VPKFILE}.sig
 	fi
 	SIGBINFILE=${VPKFILE}.sig.bin
-	cat ${VPKFILE} ${VPKFILE}.sig					> ${SIGBINFILE}
-	hl_get_size ${SIGBINFILE} 4
+	cat ${STAGE_DIR}/${VPKFILE} ${STAGE_DIR}/${VPKFILE}.sig	> ${STAGE_DIR}/${SIGBINFILE}
+	hl_get_size ${STAGE_DIR}/${SIGBINFILE} 4
 	FULLBINFILE=${SIGBINFILE}.full
-	cat ${SIGBINFILE}.size ${SIGBINFILE}				> ${FULLBINFILE}
+	cat ${STAGE_DIR}/${SIGBINFILE}.size ${STAGE_DIR}/${SIGBINFILE}	> ${STAGE_DIR}/${FULLBINFILE}
 	PADFULLBINFILE=${FULLBINFILE}.pad
-	dd status=none conv=sync bs=16 if=${FULLBINFILE}		> ${PADFULLBINFILE}
+	dd status=none conv=sync bs=16 if=${STAGE_DIR}/${FULLBINFILE}	> ${STAGE_DIR}/${PADFULLBINFILE}
 
-	SIZE_WITHOUT_CRC=$(stat --printf="%s" ${PADFULLBINFILE})
+	SIZE_WITHOUT_CRC=$(stat --printf="%s" ${STAGE_DIR}/${PADFULLBINFILE})
 	NUM_SLICES=$((SIZE_WITHOUT_CRC/MAX_SPI_DMA_DATA_TRANSFER))
 	MODULO_SLICES=$((SIZE_WITHOUT_CRC%MAX_SPI_DMA_DATA_TRANSFER))
 	if [ ${MODULO_SLICES} -ne 0 ]; then
@@ -181,14 +184,14 @@
 
 	# add 4 bytes for each CRCs
 	SIZE_CRCS=$((NUM_SLICES*4))
-	hl_get_size ${FULLBINFILE} ${SIZE_CRCS}
-	dd status=none conv=notrunc seek=0 bs=${SIZE_CRCS} if=${FULLBINFILE}.size of=${PADFULLBINFILE}
+	hl_get_size ${STAGE_DIR}/${FULLBINFILE} ${SIZE_CRCS}
+	dd status=none conv=notrunc seek=0 bs=${SIZE_CRCS} if=${STAGE_DIR}/${FULLBINFILE}.size of=${STAGE_DIR}/${PADFULLBINFILE}
 	local leftover=${SIZE_WITHOUT_CRC}
 	local skip_offset=0
 
 	for ((slice_cnt=0; slice_cnt<${NUM_SLICES}; slice_cnt++)); do
 		# bin file: (64KB-16) + CRC
-		local BIN_FILE=${PADFULLBINFILE}.part${slice_cnt}
+		local BIN_FILE=${STAGE_DIR}/${PADFULLBINFILE}.part${slice_cnt}
 		if [ $leftover -lt $MAX_SPI_DMA_DATA_TRANSFER ]; then
 			size_bytes=$leftover
 		else
@@ -196,19 +199,19 @@
 		fi
 
 		size_words=$((size_bytes/4))
-		dd status=none skip=${skip_offset} conv=sync bs=4 if=${PADFULLBINFILE} count=${size_words} > ${BIN_FILE}
+		dd status=none skip=${skip_offset} conv=sync bs=4 if=${STAGE_DIR}/${PADFULLBINFILE} count=${size_words} > ${BIN_FILE}
 		hl_calc_crc32 ${BIN_FILE}
 		if [ ${CRC_ERROR} -eq 0 ]; then
 			cat ${BIN_FILE}.crc32 >> ${BIN_FILE}
 		fi
 		flash_offset=$((SPI_DMA_DEST_64KB_OFFS/4))
 		flash_offset=$((flash_offset*slice_cnt))
-		dd status=none conv=sync if=${BIN_FILE} bs=4 seek=$flash_offset of=${DEV_PREFIX}-${FILEPATH}.le
+		dd status=none conv=sync if=${BIN_FILE} bs=4 seek=$flash_offset of=${BINS_DIR}/${DEV_PREFIX}-${FILENAME}.le
 		leftover=$((leftover-size_bytes))
 		skip_offset=$((skip_offset+size_words))
 	done
-	hl_pldm_hex $(basename ${FILEPATH}) ${SIGBINFILE}.full 16
-	rm -f ${FUNCNAME}-*
+	hl_pldm_hex ${FILENAME} ${STAGE_DIR}/${SIGBINFILE}.full 16 ${HEXS_DIR}
+	rm -f ${STAGE_DIR}/${FUNCNAME}-*
 }
 
 function calculate_hash {
@@ -221,7 +224,7 @@
 function get_image_size {
 	local name=${1^^}
 
-	stat -Lc "%s" $(eval "echo ${DEV_PREFIX}-\${${name}_BINFILE}")
+	stat -Lc "%s" $(eval "echo ${BINS_DIR}/${DEV_PREFIX}-\${${name}_BINFILE}")
 }
 
 function hl_join_bin {
@@ -382,11 +385,13 @@
 
 function prep_erom_with_crc {
 	BINFILE=$1
-	cat ${BINFILE}>${BINFILE}.tmp.pad
-	dd status=none conv=sync bs=16 if=${BINFILE}.tmp.pad of=${BINFILE}.tmp.algn
-	hl_calc_crc32 ${BINFILE}.tmp.algn
-	cat ${BINFILE}.tmp.algn ${BINFILE}.tmp.algn.crc32 > ${DEV_PREFIX}-${BINFILE}
-	rm -rf ${BINFILE}.tmp*
+	local BINFILE_BASE=$(basename ${BINFILE})
+
+	cat ${BINFILE} > ${STAGE_DIR}/${BINFILE_BASE}.tmp.pad
+	dd status=none conv=sync bs=16 if=${STAGE_DIR}/${BINFILE_BASE}.tmp.pad of=${STAGE_DIR}/${BINFILE_BASE}.tmp.algn
+	hl_calc_crc32 ${STAGE_DIR}/${BINFILE_BASE}.tmp.algn
+	cat ${STAGE_DIR}/${BINFILE_BASE}.tmp.algn ${STAGE_DIR}/${BINFILE_BASE}.tmp.algn.crc32 > ${BINS_DIR}/${DEV_PREFIX}-${BINFILE_BASE}
+	rm -rf ${STAGE_DIR}/${BINFILE_BASE}.tmp*
 }
 
 function prep_erom_fit {
@@ -394,48 +399,48 @@
 
 	ITS_SED_PARAMS+="s/@HL_DEV@/${DEV_PREFIX}/g;"
 	ITS_SED_PARAMS+="s|@HL_EROM_VER_FILE@|${EROM_VERSION_FILE}|g;"
-	sed ${ITS_SED_PARAMS} ${EROM_ITS_TEMPLATE} > ${DEV_PREFIX}-erom.its
-	$MKIMAGE_FILE -f ${DEV_PREFIX}-erom.its ${ITB_FILE}
+	sed ${ITS_SED_PARAMS} ${EROM_ITS_TEMPLATE} > ${STAGE_DIR}/${DEV_PREFIX}-erom.its
+	$MKIMAGE_FILE -f ${STAGE_DIR}/${DEV_PREFIX}-erom.its ${ITB_FILE}
 }
 
 function prep_image_erom {
 	local OUTF=$1
 	local HI_OFFSET=""
 
-	prep_erom_with_crc ${EROM_PCIE_BINFILE}
-	prep_erom_with_crc ${EROM_RECOVERY_BINFILE}
+	prep_erom_with_crc ${STAGE_DIR}/${EROM_PCIE_BINFILE}
+	prep_erom_with_crc ${STAGE_DIR}/${EROM_RECOVERY_BINFILE}
 
 	# create fit file with erom additional data
-	prep_erom_fit ${EROM_FIT_BINFILE}
-	prep_erom_with_crc ${EROM_FIT_BINFILE}
+	prep_erom_fit ${ITBS_DIR}/${EROM_FIT_BINFILE}
+	prep_erom_with_crc ${ITBS_DIR}/${EROM_FIT_BINFILE}
 
 	# Create erom image pointers binary
-	prep_image_ptrs_erom ${EROM_IMAGES_POINTERS_BINFILE}
-	prep_erom_with_crc ${EROM_IMAGES_POINTERS_BINFILE}
+	prep_image_ptrs_erom ${BINS_DIR}/${EROM_IMAGES_POINTERS_BINFILE}
+	prep_erom_with_crc ${BINS_DIR}/${EROM_IMAGES_POINTERS_BINFILE}
 
 	if [ ${HL_TARGET} == "pldm" ]; then
 		#Create erom hex files for paladium
-		hl_pldm_flash_hex erom-binary-pcie ${DEV_PREFIX}-${EROM_PCIE_BINFILE}
-		hl_pldm_flash_hex erom-binary-recovery ${DEV_PREFIX}-${EROM_RECOVERY_BINFILE}
+
+		hl_pldm_flash_hex erom-binary-pcie ${BINS_DIR}/${DEV_PREFIX}-${EROM_PCIE_BINFILE} ${HEXS_DIR}
+		hl_pldm_flash_hex erom-binary-recovery ${BINS_DIR}/${DEV_PREFIX}-${EROM_RECOVERY_BINFILE} ${HEXS_DIR}
 
 		#Create erom image pointers hex files for paladium
-		hl_pldm_flash_hex erom-binary-img-ptr ${DEV_PREFIX}-${EROM_IMAGES_POINTERS_BINFILE}
+		hl_pldm_flash_hex erom-binary-img-ptr ${BINS_DIR}/${DEV_PREFIX}-${EROM_IMAGES_POINTERS_BINFILE} ${HEXS_DIR}
 	fi
 
 	# create little-endian binaries for ASIC / ONIX
-	cp -f ${DEV_PREFIX}-${EROM_PCIE_BINFILE} ${DEV_PREFIX}-${EROM_PCIE_BINFILE}.le
-	cp -f ${DEV_PREFIX}-${EROM_RECOVERY_BINFILE} ${DEV_PREFIX}-${EROM_RECOVERY_BINFILE}.le
-	cp -f ${DEV_PREFIX}-${EROM_FIT_BINFILE} ${DEV_PREFIX}-${EROM_FIT_BINFILE}.le
-	cp -f ${DEV_PREFIX}-${EROM_IMAGES_POINTERS_BINFILE} ${DEV_PREFIX}-${EROM_IMAGES_POINTERS_BINFILE}.le
-
+	cp -f ${BINS_DIR}/${DEV_PREFIX}-${EROM_PCIE_BINFILE} ${BINS_DIR}/${DEV_PREFIX}-${EROM_PCIE_BINFILE}.le
+	cp -f ${BINS_DIR}/${DEV_PREFIX}-${EROM_RECOVERY_BINFILE} ${BINS_DIR}/${DEV_PREFIX}-${EROM_RECOVERY_BINFILE}.le
+	cp -f ${BINS_DIR}/${DEV_PREFIX}-${EROM_FIT_BINFILE} ${BINS_DIR}/${DEV_PREFIX}-${EROM_FIT_BINFILE}.le
+	cp -f ${BINS_DIR}/${DEV_PREFIX}-${EROM_IMAGES_POINTERS_BINFILE} ${BINS_DIR}/${DEV_PREFIX}-${EROM_IMAGES_POINTERS_BINFILE}.le
 	# Reserved space (80820 bytes) filled by 0xFF
 	HI_OFFSET=$(printf %d ${EROM_TPM_OFFSET})
 	dd status=none conv=sync bs=${HI_OFFSET} count=1 if=/dev/zero |  tr "\000" "\377" > ${OUTF}
 
-	dd status=none seek=$(printf %d ${EROM_START_OFFSET})		if=${DEV_PREFIX}-${EROM_IMAGES_POINTERS_BINFILE}.le	bs=1 of=${OUTF}
-	dd status=none seek=$(printf %d ${EROM_PCIE_OFFSET})		if=${DEV_PREFIX}-${EROM_PCIE_BINFILE}.le		bs=1 of=${OUTF}
-	dd status=none seek=$(printf %d ${EROM_RECOVERY_OFFSET})	if=${DEV_PREFIX}-${EROM_RECOVERY_BINFILE}.le		bs=1 of=${OUTF}
-	dd status=none seek=$(printf %d ${EROM_FIT_OFFSET})		if=${DEV_PREFIX}-${EROM_FIT_BINFILE}.le			bs=1 of=${OUTF}
+	dd status=none seek=$(printf %d ${EROM_START_OFFSET})		if=${BINS_DIR}/${DEV_PREFIX}-${EROM_IMAGES_POINTERS_BINFILE}.le	bs=1 of=${OUTF}
+	dd status=none seek=$(printf %d ${EROM_PCIE_OFFSET})		if=${BINS_DIR}/${DEV_PREFIX}-${EROM_PCIE_BINFILE}.le		bs=1 of=${OUTF}
+	dd status=none seek=$(printf %d ${EROM_RECOVERY_OFFSET})	if=${BINS_DIR}/${DEV_PREFIX}-${EROM_RECOVERY_BINFILE}.le	bs=1 of=${OUTF}
+	dd status=none seek=$(printf %d ${EROM_FIT_OFFSET})		if=${BINS_DIR}/${DEV_PREFIX}-${EROM_FIT_BINFILE}.le		bs=1 of=${OUTF}
 
 }
 
@@ -495,15 +500,15 @@
 	local PRIME_IMAGE_RAM_CRC_ERROR=$3
 	local PRIME_IMAGE_RAM_SIGN_ERROR=$4
 
-	dd if=/dev/zero bs=1 status=none count=1016 	> preboot-pad.bin
-	cat ${PREBOOT_BINFILE} 				>> preboot-pad.bin
-	mv ${PREBOOT_BINFILE} preboot-orig.bin
-	mv preboot-pad.bin ${PREBOOT_BINFILE}
+	dd if=/dev/zero bs=1 status=none count=1016 	> ${STAGE_DIR}/preboot-pad.bin
+	cat ${PREBOOT_BINFILE} 				>> ${STAGE_DIR}/preboot-pad.bin
+	mv ${PREBOOT_BINFILE} ${STAGE_DIR}/preboot-orig.bin
+	mv ${STAGE_DIR}/preboot-pad.bin ${PREBOOT_BINFILE}
 
 	hl_pack_bin ${PREBOOT_BINFILE} 64K ${PRIME_IMAGE_RAM_KEY} ${PRIME_IMAGE_RAM_CRC_ERROR} ${PRIME_IMAGE_RAM_SIGN_ERROR}
 
 	rm -rf ${PREBOOT_BINFILE}
-	mv preboot-orig.bin ${PREBOOT_BINFILE}
+	mv ${STAGE_DIR}/preboot-orig.bin ${PREBOOT_BINFILE}
 }
 
 function zephyr_preboot_d1_pack {
diff --git a/board/habanalabs/gaudi3/post-image-fit-descriptor.sh b/board/habanalabs/gaudi3/post-image-fit-descriptor.sh
index 5a5711d..97109ba 100644
--- a/board/habanalabs/gaudi3/post-image-fit-descriptor.sh
+++ b/board/habanalabs/gaudi3/post-image-fit-descriptor.sh
@@ -2,6 +2,6 @@
 
 function hl_handle_desc
 {
-	openssl dgst -sha384 -binary ${HL_DEV}-boot-fit.itb > img-hash-${HL_DEV}-boot-fit.sha384
-	add_descriptor_to_fit ${HL_DEV}-boot-fit.itb
+	openssl dgst -sha384 -binary ${ITBS_DIR}${HL_DEV}-boot-fit.itb > ${ITBS_DIR}img-hash-${HL_DEV}-boot-fit.sha384
+	add_descriptor_to_fit ${ITBS_DIR}${HL_DEV}-boot-fit.itb
 }
diff --git a/board/habanalabs/gaudi3/post-image-fit.sh b/board/habanalabs/gaudi3/post-image-fit.sh
index 9bcfd6b..5f7b585 100644
--- a/board/habanalabs/gaudi3/post-image-fit.sh
+++ b/board/habanalabs/gaudi3/post-image-fit.sh
@@ -8,7 +8,7 @@
 
 ETH_PHY_FW_HEX_FILE=${DEV_BASE}/eth_phy_fw.hex
 ETH_PHY_FW_HASH_FILE=${DEV_BASE}/eth_phy_fw.md5
-ETH_PHY_FW_BIN_FILE=${BR2_EXTERNAL}/output/${DEV_NAME}_arc/images/${DEV_NAME}-eth-phy.bin.le
+ETH_PHY_FW_BIN_FILE=${BINS_DIR}/${DEV_NAME}-eth-phy.bin.le
 
 hl_create_section_variables
 
@@ -52,12 +52,12 @@
 	FILEPATH=$1
 
 	# copy binaries from other HL targets required for boot fit creation
-	cp ${BR2_EXTERNAL}/output/${DEV_NAME}_arc/images/${DEV_PREFIX}-zephyr-pid-iccm.bin ${FILEPATH}/${DEV_PREFIX}-zephyr-pid-iccm.bin.le
-	cp ${BR2_EXTERNAL}/output/${DEV_NAME}_arc/images/${DEV_PREFIX}-zephyr-pid-dccm.bin ${FILEPATH}/${DEV_PREFIX}-zephyr-pid-dccm.bin.le
-	cp ${BR2_EXTERNAL}/output/${DEV_NAME}_arc/images/${DEV_PREFIX}-zephyr-mgmt-iccm.bin ${FILEPATH}/${DEV_PREFIX}-zephyr-mgmt-iccm.bin.le
-	cp ${BR2_EXTERNAL}/output/${DEV_NAME}_arc/images/${DEV_PREFIX}-zephyr-mgmt-dccm.bin ${FILEPATH}/${DEV_PREFIX}-zephyr-mgmt-dccm.bin.le
-	cp ${BR2_EXTERNAL}/output/${DEV_NAME}_arc/images/${DEV_PREFIX}-zephyr-event-iccm.bin ${FILEPATH}/${DEV_PREFIX}-zephyr-event-iccm.bin.le
-	cp ${BR2_EXTERNAL}/output/${DEV_NAME}_arc/images/${DEV_PREFIX}-zephyr-event-dccm.bin ${FILEPATH}/${DEV_PREFIX}-zephyr-event-dccm.bin.le
+	cp ${STAGE_DIR}/${DEV_PREFIX}-zephyr-pid-iccm.bin ${FILEPATH}/${DEV_PREFIX}-zephyr-pid-iccm.bin.le
+	cp ${STAGE_DIR}/${DEV_PREFIX}-zephyr-pid-dccm.bin ${FILEPATH}/${DEV_PREFIX}-zephyr-pid-dccm.bin.le
+	cp ${STAGE_DIR}/${DEV_PREFIX}-zephyr-mgmt-iccm.bin ${FILEPATH}/${DEV_PREFIX}-zephyr-mgmt-iccm.bin.le
+	cp ${STAGE_DIR}/${DEV_PREFIX}-zephyr-mgmt-dccm.bin ${FILEPATH}/${DEV_PREFIX}-zephyr-mgmt-dccm.bin.le
+	cp ${STAGE_DIR}/${DEV_PREFIX}-zephyr-event-iccm.bin ${FILEPATH}/${DEV_PREFIX}-zephyr-event-iccm.bin.le
+	cp ${STAGE_DIR}/${DEV_PREFIX}-zephyr-event-dccm.bin ${FILEPATH}/${DEV_PREFIX}-zephyr-event-dccm.bin.le
 
 	prep_eth_phy_fw_binary ${ETH_PHY_FW_HEX_FILE} ${ETH_PHY_FW_BIN_FILE}
 
diff --git a/board/habanalabs/gaudi3_arc/post-image-spi.sh b/board/habanalabs/gaudi3_arc/post-image-spi.sh
index 76bc3c4..e93380a 100755
--- a/board/habanalabs/gaudi3_arc/post-image-spi.sh
+++ b/board/habanalabs/gaudi3_arc/post-image-spi.sh
@@ -68,7 +68,7 @@
 
 function prep_image_fake_fuse {
 	local OUTF=$1
-	local TMPNAME=fake_fuse.bin
+	local TMPNAME=${STAGE_DIR}/fake_fuse.bin
 
 	# Master die fuse content
 	#Bank 0
@@ -98,7 +98,7 @@
 	PROP_NAME=$2
 	OUT_NAME=$3
 
-	TMPNAME=${KEY_NAME}-${PROP_NAME}.txt
+	TMPNAME=${STAGE_DIR}/${KEY_NAME}-${PROP_NAME}.txt
 	fdtget ${KEY_NAME}.dtb /signature/key-${KEY_NAME} rsa,${PROP_NAME} -t bx > ${TMPNAME}
 	sed 's/\s\+/\n/g' ${TMPNAME} > ${TMPNAME}.tmp
 	sed -e 's/\b[0-9]\b/0&/' -e 's/\b[a-f]\b/0&/' ${TMPNAME}.tmp  > ${TMPNAME}.tmp.le
@@ -115,37 +115,39 @@
 		KEY_NAME=$(echo -n "cust${KEY_NUM}_${KEY_BASE_NAME}")
 	fi
 
-	extract_key_props ${KEY_NAME} r-squared rr.prop
+	extract_key_props ${KEY_NAME} r-squared ${STAGE_DIR}/rr.prop
 
-	extract_key_props ${KEY_NAME} modulus mod.prop
+	extract_key_props ${KEY_NAME} modulus ${STAGE_DIR}/mod.prop
 
-	extract_key_props ${KEY_NAME} exponent exp.prop
+	extract_key_props ${KEY_NAME} exponent ${STAGE_DIR}/exp.prop
 
-	extract_key_props ${KEY_NAME} n0-inverse inv.proptemp
-	swap_file inv.proptemp inv.prop
+	extract_key_props ${KEY_NAME} n0-inverse ${STAGE_DIR}/inv.proptemp
+	swap_file ${STAGE_DIR}/inv.proptemp ${STAGE_DIR}/inv.prop
 
-	extract_key_props ${KEY_NAME} num-bits nb.proptemp
-	swap_file nb.proptemp nb.prop
+	extract_key_props ${KEY_NAME} num-bits ${STAGE_DIR}/nb.proptemp
+	swap_file ${STAGE_DIR}/nb.proptemp ${STAGE_DIR}/nb.prop
 
 	#no separate public key will be part of the image, since these props serve the purpose.
 	# Hash will be computed on the Props itself
-	cat rr.prop mod.prop exp.prop inv.prop nb.prop	> ${KEY_NAME}.pub
-	cat rr.prop mod.prop exp.prop inv.prop nb.prop	> ${KEY_NAME}-key-props.bin
+	cat ${STAGE_DIR}/rr.prop ${STAGE_DIR}/mod.prop ${STAGE_DIR}/exp.prop ${STAGE_DIR}/inv.prop nb.prop	> ${STAGE_DIR}/${KEY_NAME}.pub
+	cat ${STAGE_DIR}/rr.prop ${STAGE_DIR}/mod.prop ${STAGE_DIR}/exp.prop ${STAGE_DIR}/inv.prop nb.prop	> ${STAGE_DIR}/${KEY_NAME}-key-props.bin
 
-	rm -f *.txt *.proptemp *.prop
+	rm -f ${STAGE_DIR}/*.txt ${STAGE_DIR}/*.proptemp ${STAGE_DIR}/*.prop
 }
 
 MAX_SPI_DMA_DATA_TRANSFER=65520
 SPI_DMA_DEST_64KB_OFFS=65536
 function hl_pack_bin {
-	FILEPATH=$1
+	FILEFULLPATH=$1
+	FILENAME=$(basename ${FILEFULLPATH})
+	FILEPATH=$(dirname ${FILEFULLPATH})
 	BSIZE=$2
 	KEY_NUM=$3
 	CRC_ERROR=$4
 	SIGN_ERROR=$5
 
-	PADFULLBINFILE=${FILEPATH}.pad
-	dd status=none conv=sync bs=16 if=${FILEPATH} of=${PADFULLBINFILE}
+	PADFULLBINFILE=${FILEFULLPATH}.pad
+	dd status=none conv=sync bs=16 if=${FILEFULLPATH} of=${PADFULLBINFILE}
 
 	SIZE_WITHOUT_CRC=$(stat --printf="%s" ${PADFULLBINFILE})
 	NUM_SLICES=$((SIZE_WITHOUT_CRC/MAX_SPI_DMA_DATA_TRANSFER))
@@ -196,13 +198,13 @@
 		fi
 		flash_offset=$((SPI_DMA_DEST_64KB_OFFS/4))
 		flash_offset=$((flash_offset*slice_cnt))
-		dd status=none conv=sync if=${BIN_FILE} bs=4 seek=$flash_offset of=${DEV_PREFIX}-${FILEPATH}.le.int
+		dd status=none conv=sync if=${BIN_FILE} bs=4 seek=$flash_offset of=${STAGE_DIR}/${DEV_PREFIX}-${FILENAME}.le.int
 		leftover=$((leftover-size_bytes))
 		skip_offset=$((skip_offset+size_words))
 	done
 
-	hl_reverse_words ${DEV_PREFIX}-${FILEPATH}.le.int ${DEV_PREFIX}-${FILEPATH}.le 4
-	rm -f ${DEV_PREFIX}-${FILEPATH}.le.int
+	hl_reverse_words ${STAGE_DIR}/${DEV_PREFIX}-${FILENAME}.le.int ${BINS_DIR}/${DEV_PREFIX}-${FILENAME}.le 4
+	rm -f ${STAGE_DIR}/${DEV_PREFIX}-${FILENAME}.le.int
 }
 
 function zephyr_preboot_pack {
@@ -285,7 +287,7 @@
 function get_image_size {
 	local name=${1^^}
 
-	stat -Lc "%s" $(eval "echo \${${name}_BINFILE}")
+	stat -Lc "%s" $(eval "echo ${BINS_DIR}\${${name}_BINFILE}")
 }
 
 
@@ -323,7 +325,7 @@
 	cat ${BINFILE}>${BINFILE}.tmp.pad
 	dd status=none conv=sync bs=16 if=${BINFILE}.tmp.pad of=${BINFILE}.tmp.algn
 	hl_calc_crc32 ${BINFILE}.tmp.algn
-	cat ${BINFILE}.tmp.algn ${BINFILE}.tmp.algn.crc32 > ${BINS_DIR}${DEV_PREFIX}-${OUTFILENAME}
+	cat ${BINFILE}.tmp.algn ${BINFILE}.tmp.algn.crc32 > ${BINS_DIR}/${DEV_PREFIX}-${OUTFILENAME}
 	rm -rf ${BINFILE}.tmp*
 }
 
@@ -352,16 +354,16 @@
 	local BINFILE=$(eval "echo \${${NAME}_BINFILE}")
 
 	#align the source binaries to 16 bytes
-	dd status=none conv=sync bs=16 if=${BINFILE} of=${BINFILE}.tmp.pad
+	dd status=none conv=sync bs=16 if=${BINS_DIR}/${BINFILE} of=${STAGE_DIR}/${BINFILE}.tmp.pad
 	# Prepare checksum
-	local CKSUM=`sha384sum ${BINFILE}.tmp.pad`
+	local CKSUM=`sha384sum ${STAGE_DIR}/${BINFILE}.tmp.pad`
 	hl_chksum_add ${CKSUM}
-	rm -f ${BINFILE}.tmp.pad
+	rm -f ${STAGE_DIR}/${BINFILE}.tmp.pad
 
 	if [ $img_name == "PREBOOT" -o $img_name == "PPBOOT" ]
 	then
-		local SHA384_BINFILE=img-hash-${name%.*}.sha384
-		rm -f ${SHA384_BINFILE}.sha384
+		local SHA384_BINFILE=${STAGE_DIR}/img-hash-${name%.*}.sha384
+		rm -f ${SHA384_BINFILE}
 		echo -n "$CKSUM" | xxd -r -p > ${SHA384_BINFILE}
 	fi
 }
@@ -457,8 +459,8 @@
 	local EROM_IMG=$2
 
 	# Prepare the NEXT and MODIFIABLE for backward compatibility
-	echo "11110000" > ${HL_DEV}-next-image.txt
-	dd status=none conv=sync bs=8 count=1 if=${HL_DEV}-next-image.txt of=$img_file
+	echo "11110000" > ${STAGE_DIR}/${HL_DEV}-next-image.txt
+	dd status=none conv=sync bs=8 count=1 if=${STAGE_DIR}/${HL_DEV}-next-image.txt of=$img_file
 
 	# Pointers structure magic "NXT"0x0
 	printf "%x%x%x00" "'N" "'X" "'T" | rev_bytes_stdin | xxd -r -p -s +8 - $img_file
@@ -546,11 +548,13 @@
 	hl_get_key_props ${KEY_NUM}
 
 	TEMPNAME=${FUNCNAME}
-	dd status=none conv=sync bs=16 if=$img_file			> ${TEMPNAME}
+	TEMPFULLNAME=${STAGE_DIR}/${FUNCNAME}
+	dd status=none conv=sync bs=16 if=$img_file > ${TEMPFULLNAME}
 
 	PKFILE=${TEMPNAME}.payload.key
+	PKFULLFILE=${TEMPFULLNAME}.payload.key
 
-	cat ${TEMPNAME} ${KEY_NAME}-key-props.bin > ${PKFILE}
+	cat ${TEMPFULLNAME} ${STAGE_DIR}/${KEY_NAME}-key-props.bin > ${PKFULLFILE}
 
 	#The size of the complete image(Payload + Key + Signature) needs to be calculated
 	#and placed in the manifest even before hashing and signing over it as the size
@@ -562,7 +566,7 @@
 	#Hence the final formula = ALIGN_16(Payload + Key + Signature) + CRC
 
 	#payload + key size
-	local pk_sz=$(stat -Lc "%s" $(eval "echo \${PKFILE}"))
+	local pk_sz=$(stat -Lc "%s" $(eval "echo \${PKFULLFILE}"))
 	#size to be aligned to
 	local align_sz=16
 	#payload + key + signature size
@@ -574,17 +578,17 @@
 
 	printf "%08x" $img_desc_sz | rev_bytes_stdin | xxd -r -p -s +4 - ${PKFILE}
 
-	$OPENSSL dgst -sha384 -binary ${PKFILE} > ${PKFILE}.sha384
+	$OPENSSL dgst -sha384 -binary ${PKFULLFILE} > ${PKFULLFILE}.sha384
 
-	hl_sign_bin ${PKFILE} ${KEY_NAME}
+	hl_sign_bin ${PKFULLFILE} ${STAGE_DIR}/${KEY_NAME}
 
 	SIGBINFILE=${PKFILE}.sig.bin
 
-	cat ${PKFILE} ${PKFILE}.sig			>${SIGBINFILE}
+	cat ${PKFULLFILE} ${PKFULLFILE}.sig > ${STAGE_DIR}/${SIGBINFILE}
 
-	prep_img_desc_with_crc ${SIGBINFILE}
+	prep_img_desc_with_crc ${STAGE_DIR}/${SIGBINFILE}
 
-	hl_reverse_words ${DEV_PREFIX}-${SIGBINFILE} $out_file 4
+	hl_reverse_words ${STAGE_DIR}/${DEV_PREFIX}-${SIGBINFILE} $out_file 4
 }
 
 function prep_image_ptrs {
@@ -592,81 +596,89 @@
 }
 
 function prep_pldm_preboot_d1_hex {
-	swap_file ${PREBOOT_D1_BINFILE} ${PREBOOT_D1_BINFILE}.orig.be
-	hl_pldm_hex preboot_d1 ${PREBOOT_D1_BINFILE}.orig.be 1
+	swap_file ${BINS_DIR}/${PREBOOT_D1_BINFILE} ${BINS_DIR}/${PREBOOT_D1_BINFILE}.orig.be
+	hl_pldm_hex preboot_d1 ${BINS_DIR}/${PREBOOT_D1_BINFILE}.orig.be 1 ${HEXS_DIR}
 }
 
 function prep_image_erom {
-	local OUTF=$1
+	local OUTFULLPATH=$1
+	local OUTF=$(basename $1)
 
-	dd status=none conv=sync bs=16 if=${EROM_PCIE_BINFILE} of=${EROM_PCIE_PADDED_BINFILE}
-	dd status=none conv=sync bs=16 if=${EROM_RECOVERY_BINFILE} of=${EROM_RECOVERY_PADDED_BINFILE}
+	dd status=none conv=sync bs=16 if=${BINS_DIR}/${EROM_PCIE_BINFILE} of=${BINS_DIR}/${EROM_PCIE_PADDED_BINFILE}
+	dd status=none conv=sync bs=16 if=${BINS_DIR}/${EROM_RECOVERY_BINFILE} of=${BINS_DIR}/${EROM_RECOVERY_PADDED_BINFILE}
 
-	prep_image_ptrs_common_v4 ${DEV_PREFIX}-erom-images-pointers.bin.le $IMAGE_TYPE_EROM
+	prep_image_ptrs_common_v4 ${STAGE_DIR}/${DEV_PREFIX}-erom-images-pointers.bin.le $IMAGE_TYPE_EROM
 
-	prep_img_desc_with_crc ${EROM_PCIE_BINFILE}
-	prep_img_desc_with_crc ${EROM_RECOVERY_BINFILE}
+	prep_img_desc_with_crc ${STAGE_DIR}/${EROM_PCIE_BINFILE}
+	prep_img_desc_with_crc ${STAGE_DIR}/${EROM_RECOVERY_BINFILE}
 
-	hl_reverse_words ${DEV_PREFIX}-erom-images-pointers.bin.le \
-	${DEV_PREFIX}-erom-images-pointers.bin.be 4
+	hl_reverse_words ${STAGE_DIR}/${DEV_PREFIX}-erom-images-pointers.bin.le \
+	${BINS_DIR}/${DEV_PREFIX}-erom-images-pointers.bin.be 4
 
 	if [ ${HL_TARGET} == "pldm" ]; then
 	#Create erom hex files for paladium
-		hl_pldm_flash_hex erom-binary-pcie ${DEV_PREFIX}-${EROM_PCIE_BINFILE}
-		hl_pldm_flash_hex erom-binary-recovery ${DEV_PREFIX}-${EROM_RECOVERY_BINFILE}
+		hl_pldm_flash_hex erom-binary-pcie ${BINS_DIR}/${DEV_PREFIX}-${EROM_PCIE_BINFILE} ${HEXS_DIR}
+		hl_pldm_flash_hex erom-binary-recovery ${BINS_DIR}/${DEV_PREFIX}-${EROM_RECOVERY_BINFILE} ${HEXS_DIR}
 
 		#Create erom image pointers hex files for paladium
-		hl_pldm_flash_hex erom-binary-img-ptr ${DEV_PREFIX}-erom-images-pointers.bin.be
+		hl_pldm_flash_hex erom-binary-img-ptr ${BINS_DIR}/${DEV_PREFIX}-erom-images-pointers.bin.be ${HEXS_DIR}
 	fi
 
-	dd status=none conv=sync if=${DEV_PREFIX}-erom-images-pointers.bin.${ENDIANNESS} \
-	bs=$(printf %d ${EROM_PCIE_OFFSET})	> ${OUTF}.tmp
+	dd status=none conv=sync if=${BINS_DIR}/${DEV_PREFIX}-erom-images-pointers.bin.${ENDIANNESS} \
+	bs=$(printf %d ${EROM_PCIE_OFFSET})	> ${STAGE_DIR}/${OUTF}.tmp
 	dd status=none conv=sync seek=$(printf %d ${EROM_PCIE_OFFSET}) \
-	if=${DEV_PREFIX}-${EROM_PCIE_BINFILE} bs=1 of=${OUTF}.tmp
+	if=${BINS_DIR}/${DEV_PREFIX}-${EROM_PCIE_BINFILE} bs=1 of=${STAGE_DIR}/${OUTF}.tmp
 	dd status=none conv=sync seek=$(printf %d ${EROM_RECOVERY_OFFSET}) \
-	if=${DEV_PREFIX}-${EROM_RECOVERY_BINFILE} bs=1 of=${OUTF}.tmp
+	if=${BINS_DIR}/${DEV_PREFIX}-${EROM_RECOVERY_BINFILE} bs=1 of=${STAGE_DIR}/${OUTF}.tmp
 
-	hl_reverse_words ${OUTF}.tmp ${OUTF} 4
-	rm -f ${OUTF}.tmp
+	hl_reverse_words ${STAGE_DIR}/${OUTF}.tmp ${OUTFULLPATH} 4
+	rm -f ${STAGE_DIR}/${OUTF}.tmp
 }
 
 function hl_objcopy_sections() {
-	local PREFIX=$1
+	local FILEFULLPATH=$1
+	local FILENAME=$(basename ${FILEFULLPATH})
+	local FILEPATH=$(dirname ${FILEFULLPATH})
+	#local EXT=${FILENAME##*.}
+	local PREFIX=${FILENAME%.*}
 	local ALIGN=$2
 	local MEM=$3
 	local SECTIONS=${*:4}
 
 	${OBJCOPY} -S --gap-fill 0xff --output-target=binary --remove-section=.comment\
-	$(echo -n $SECTIONS | sed 's/[^ ]* */ --only-section=&/g') ${STAGE_DIR}/${PREFIX}.elf \
+	$(echo -n $SECTIONS | sed 's/[^ ]* */ --only-section=&/g') ${FILEPATH}/${PREFIX}.elf \
 	--set-section-flags=bss=alloc,load,contents --set-section-flags=.stack=alloc,load,contents\
 	${STAGE_DIR}/${PREFIX}-${MEM}.bin.int
 
 	dd status=none conv=sync bs=${ALIGN} if=${STAGE_DIR}/${PREFIX}-${MEM}.bin.int > \
-	${STAGE_DIR}/${PREFIX}-${MEM}.bin
+	${BINS_DIR}/${PREFIX}-${MEM}.bin
 
 	rm -f ${STAGE_DIR}/${PREFIX}-${MEM}.bin.int
 }
 
 function prep_image_pcie_pif {
-	local PIF_FILENAME=$1
-	local PREFIX=$(echo "$1" | cut -f 1 -d '.')
+	local FILEFULLPATH=$1
+	local FILENAME=$(basename ${FILEFULLPATH})
+	local FILEPATH=$(dirname ${FILEFULLPATH})
+	local PREFIX=${FILENAME%.*}
+	#local EXT=${FILENAME##*.}
+	local PREFIX=${FILENAME%.*}
 	local ICCM_SECTIONS=(.ivt .text)
 	local DCCM_SECTIONS=(.data .sdata .bss .stack)
 
 	# Align ICCM and DCCM binary to 16 bytes to comply with CCM ECC data bank size
-	hl_objcopy_sections ${PREFIX} 16 "iccm" ${ICCM_SECTIONS[@]}
-	hl_objcopy_sections ${PREFIX} 16 "dccm" ${DCCM_SECTIONS[@]}
+	hl_objcopy_sections ${FILEFULLPATH} 16 "iccm" ${ICCM_SECTIONS[@]}
+	hl_objcopy_sections ${FILEFULLPATH} 16 "dccm" ${DCCM_SECTIONS[@]}
 
-	hl_pack_bin ${PREFIX}-iccm.bin 4k 0 0 0
-	hl_pack_bin ${PREFIX}-dccm.bin 4k 0 0 0
-
-	swap_file ${DEV_PREFIX}-${PIF_ICCM_BINFILE}.le ${DEV_PREFIX}-${PIF_ICCM_BINFILE}.be
-	swap_file ${DEV_PREFIX}-${PIF_DCCM_BINFILE}.le ${DEV_PREFIX}-${PIF_DCCM_BINFILE}.be
+	hl_pack_bin ${BINS_DIR}/${PREFIX}-iccm.bin 4k 0 0 0
+	hl_pack_bin ${BINS_DIR}/${PREFIX}-dccm.bin 4k 0 0 0
+	swap_file ${BINS_DIR}/${DEV_PREFIX}-${PIF_ICCM_BINFILE}.le ${BINS_DIR}/${DEV_PREFIX}-${PIF_ICCM_BINFILE}.be
+	swap_file ${BINS_DIR}/${DEV_PREFIX}-${PIF_DCCM_BINFILE}.le ${BINS_DIR}/${DEV_PREFIX}-${PIF_DCCM_BINFILE}.be
 }
 
 function hl_pad_arcboot_rom {
-	mv ${ARCBOOT_BINFILE} ${ARCBOOT_BINFILE_ORIG}
+	mv ${BINS_DIR}/${ARCBOOT_BINFILE} ${BINS_DIR}/${ARCBOOT_BINFILE_ORIG}
 
 	# pad arcboot rom binary to size of ROM memory
-	dd status=none conv=sync bs=128K if=${ARCBOOT_BINFILE_ORIG} > ${ARCBOOT_BINFILE}
+	dd status=none conv=sync bs=128K if=${BINS_DIR}/${ARCBOOT_BINFILE_ORIG} > ${BINS_DIR}/${ARCBOOT_BINFILE}
 }
